---
title: "R Notebook"
author: paulbuckley
date: 14/12/2021
output: html_document
---

```{r}
library(data.table)
library(dplyr)
library(tidyr)
library(ggpubr)
library(Biostrings)
library(tidyverse)
library(foreach)
library(stringdist)
library(doParallel)

```
## Useful functions

```{r}

# Function to provide a closest match. Used to match HLA Alleles across mixed output styles.
ClosestMatch2 = function(string, stringVector){

  stringVector[amatch(string, stringVector, maxDist=Inf)]

}

```

## Aims
Question: - What Wuhan Hu-1 SARS-CoV2 T cell targets are mutated in Omicron?

# Gather data

- Gather a comprehensive dataset of immunogenic CoV2 peptides
- As of Dec 2021. By then, most of Wuhan proteome had been scanned for immunogenic CD8+ T cell targets
## IEDB + VIPR

```{r}
# Read pre-processed dataset of SARS2 immunogenic targets
IEDB_VIPR = readRDS("DATA/SARS_COV2_DATA_DEC21/DEC_2021_Human_Cleaned_SarsCoV2_EpitopeDataset_w_Freq_ONLYPOSITIVE.rds")
IEDB_VIPR_Full = IEDB_VIPR
IEDB_VIPR %>% DT::datatable()

IEDB_VIPR=IEDB_VIPR %>% select(Peptide) %>% mutate(Dataset = "IEDB+VIPR")

```

## Read immunognic epitopes from the MIRA dataset
- Only take those not observed in IEDB+VIPR
- Assumed immunogenic as confirmed recognised by TCR
```{r}

MIRA_DATA_FULL = fread(cmd = 'unzip -p MIRA_DATA_FULL.txt.zip')
MIRA_MISSING=MIRA_DATA_FULL %>% select(Peptide)%>% distinct() %>% mutate(Dataset ="MIRA") %>% filter(! Peptide %in% IEDB_VIPR$Peptide)

MIRA_MISSING %>% head()
IEDB_VIPR_Full%>% head
```

## Join MIRA + IEDB+VIPR

```{r}
# Impute missing columns for MIRA.
MIRA_MISSING_Full=MIRA_MISSING%>% mutate(HLA_Allele="Unknown", Protein_Name="Unknown", Qualitative_Measure = "Positive", Measurement = "MIRA",Immunogenicity_Binary="Positive",n=NA,Frequency_of_ImmunogenicityBinary=NA)
IEDB_VIPR_Full=IEDB_VIPR_Full %>% mutate(Dataset = "IEDB+VIPR")

IEDB_VIPR_MIRA_FULL=IEDB_VIPR_Full %>% mutate(Dataset = "IEDB+VIPR") %>% rbind(MIRA_MISSING_Full)
saveRDS(IEDB_VIPR_MIRA_FULL, file="IEDB_VIPR_MIRA_DATA_FULL_HLA.rds")
FullDataset = rbind(IEDB_VIPR,MIRA_MISSING)

saveRDS(FullDataset, "IEDB_VIPR_MIRA_COV2_DATASET.rds")

# filter only class I by length <15
FullDataset=FullDataset %>% mutate(Length = nchar(Peptide)) %>% filter(Length < 15)

```

# Map epitopes to Wuhan isolate proteome
- Find positions in Wuhan
- This is to confirm target is from Wuhan
```{r}
#list peptides
full_peptides=FullDataset %>% select(Peptide) %>% unique %>% pull

#pull in CoV2 proteome (Wuhan)
CoV2_stringset=readAAStringSet("DATA/Coronavirus_Sequences/SARS-CoV-2-Wuhan.txt",use.names = TRUE)
# For each SARS2 Wuhan epitope, find and store its location in wuhan proteome
Ep_Coverage = foreach(i = 1:length(full_peptides),.combine = rbind, .packages = c("dplyr","magrittr","data.table","Biostrings"))%do% {
  peptide_i=AAString(full_peptides[i])

  DT=as.data.table(unlist(vmatchPattern(peptide_i,CoV2_stringset)))%>% mutate(Peptide = toString(peptide_i))
  DT

}


```

```{r}
# Clean the 'Ep coverage' output. Extract the protein name
Ep_Coverage=Ep_Coverage %>% mutate(Protein = gsub(".*protein=(.+)\\].*","\\1",names))%>% mutate(Protein = gsub("\\].*","",Protein))
Ep_Coverage=Ep_Coverage %>% dplyr::rename(start_pos="start",end_pos="end")
# Create a column containing each position that each epitope exists in
Protein_positions=Ep_Coverage %>% group_by(r=row_number()) %>% mutate(allpos = list(start_pos:end_pos)) %>% unnest() %>% group_by(start_pos,end_pos) %>% dplyr::summarise(Positions = paste0(allpos,collapse = ","))%>% ungroup
WUHAN_COV2_Ep_Coverage=Ep_Coverage %>% inner_join(Protein_positions)
# Map to protein name to protein IDs
PROT_IDS=names(CoV2_stringset) %>% as.data.table() %>% dplyr::rename(names=".") %>% mutate(RefSeqProteinID = row_number())%>% mutate(Protein = gsub(".*protein=(.+)\\].*","\\1",names))%>% mutate(Protein = gsub("\\].*","",Protein))

WUHAN_COV2_Ep_Coverage=WUHAN_COV2_Ep_Coverage %>% inner_join(PROT_IDS)

```

```{r}
WUHAN_COV2_Ep_Coverage%>% select(Peptide) %>% distinct()%>% nrow
WUHAN_COV2_Ep_Coverage %>% select(Peptide, Protein) %>% distinct() %>% nrow
#saveRDS(WUHAN_COV2_Ep_Coverage,file="WUHAN_COV2_Ep_Coverage.rds")

```



```{r,dpi=300}

CORONAVIRUSSEQS = fread("DATA/Coronavirus_sequences_Full_DT.txt")%>% filter(Virus %in% 'SARS-CoV-2-Wuhan')
CORONAVIRUSSEQS=CORONAVIRUSSEQS %>% mutate(Length = nchar(`Seq.V1`)) %>% select(!c(Virus, `Seq.V1`))

```


# Map epitopes to the above regions of Omicron
## Import variant (Omicron) proteome


```{r}
# Deal w/ proteinIDs which may differ between RefSeq (Wuhan) and Variant (Omicron)
Variant_StringSet=readAAStringSet("DATA/Omicron_ProteinSeqs/Omicron_SARSCoV2_Seq.fasta",use.names = TRUE)
## Clean the name, give each protein an ID.
Variant_ProteinIDs = names(Variant_StringSet) %>% as.data.table() %>% dplyr::rename(Variantnames=".") %>%
  mutate(VariantProteinID = row_number())%>% mutate(Protein = gsub(".*protein=(.+)\\].*","\\1",Variantnames))%>%
  mutate(Protein = gsub("\\].*","",Protein))

```

```{r}
# To map refseq protein name and mutant protein name, link the ids,.
WUHAN_COV2_Ep_Coverage=WUHAN_COV2_Ep_Coverage %>% inner_join(Variant_ProteinIDs)

```


## Perform a local global to entire Variant
- For each epitope of interest, scan the mutant proteome to see if we find epitope or a mutant.
-
```{r}
# DT mapping vartiant protein name and variant protein id.
VARIANT_PROTEINS = WUHAN_COV2_Ep_Coverage %>% select(Protein, VariantProteinID) %>% distinct()

cores=detectCores()
cl <- makeCluster(cores[1]-1) #not to overload your computer
registerDoParallel(cl)
# Nested foreach: for each epitope, run throiugh each variant protein sequence.
Mutations_Epitopes = foreach(i = 1:nrow(WUHAN_COV2_Ep_Coverage),.combine = rbind, .packages = c("dplyr","magrittr","data.table","Biostrings")) %:%
foreach(ii = 1:length(Variant_StringSet),.combine = rbind, .packages = c("dplyr","magrittr","data.table","Biostrings"))%dopar% {

  peptide_i=WUHAN_COV2_Ep_Coverage$Peptide[i]
  AlignmentDT = NULL
  ALIGNMENT=pairwiseAlignment(Variant_StringSet[[ii]], peptide_i, type="local-global")
  start_match = start(pattern(ALIGNMENT))

  FullAlignmentDT = data.table(Peptide = peptide_i, #original wuhan peptide
                               VariantAlignment =   as.character(aligned(pattern(ALIGNMENT))), #best Match found in protein
                               score=score(ALIGNMENT), # alignment score
                               TestedProtein=VARIANT_PROTEINS %>% filter(VariantProteinID == ii) %>% pull(Protein), # the protein source of highest match
                               VariantProteinID=VARIANT_PROTEINS %>% filter(VariantProteinID == ii) %>% pull(VariantProteinID), #Variant protein id
                               start_match = start_match )

}

stopCluster(cl)
# Take the highest match score for each peptide. This is their match in the proteome.
Mutations_Epitopes = Mutations_Epitopes%>% distinct() %>% group_by(Peptide) %>% slice_max(score, with_ties = TRUE)%>% ungroup

Mutations_Epitopes=Mutations_Epitopes %>% dplyr::rename(Protein=TestedProtein) %>% select(Peptide, VariantAlignment, VariantProteinID, Protein,start_match)
Mutations_Epitopes = Mutations_Epitopes %>% distinct() %>% ungroup

```


```{r}

# Calculate hamming distance between WT and MT
Mutations_Epitopes=Mutations_Epitopes %>% mutate(Hamming = stringdist::stringdist(Peptide, VariantAlignment, method="hamming"))

# Join with Ep coverage data. Joins by protein. This therefore excludes any matches between WT and MT that are not to the same protein as they highly unlikely to be accurate matches.
MutationData_Wuhan_vs_Variant=WUHAN_COV2_Ep_Coverage %>% inner_join(Mutations_Epitopes)

```



```{r}
# Get rid of epitopes with H0, i.e, identical WT-MT
MUTATIONDATA=MutationData_Wuhan_vs_Variant %>% filter(! Hamming==0)
MUTATIONDATA=MUTATIONDATA %>% mutate(ProportionMutated = Hamming/nchar(Peptide))
# Function to calculate the position of the mutation
calculateMutation_position_V2 <- function(str1, str2,start_pos) {
  require(Biostrings)
  peptide = str1
  variantalignment = str2
  alignment = pairwiseAlignment(variantalignment,peptide)
  str1=as.character(aligned(pattern(alignment)))
  str2=as.character(aligned(subject(alignment)))



  str1vec <- unlist(strsplit(str1, ""))
  str2vec <- unlist(strsplit(str2, ""))
  iMut <- (1:length(str1vec))[str1vec != str2vec]
  MUTPOS = (start_pos-1)+iMut
  return(paste0(MUTPOS))
}

calculateMutation_position_V2("SSRGTSPAR","SSKRTSPAR",201)



```






```{r}
# Calculate mutation positions for each found mutant between wuhan and Omicron BA1
MUTATIONS = foreach(i = 1:nrow(MUTATIONDATA),.combine = rbind, .packages = c("dplyr","magrittr","data.table","Biostrings"))%do% {

  mutations = paste0(calculateMutation_position_V2(MUTATIONDATA$Peptide[i],MUTATIONDATA$VariantAlignment[i],MUTATIONDATA$start_pos[i]),collapse = ",")
  data.table(Peptide = MUTATIONDATA$Peptide[i],
             VariantAlignment=MUTATIONDATA$VariantAlignment[i],
             start_pos= MUTATIONDATA$start_pos[i],
             Protein = MUTATIONDATA$Protein[i],
             MutationPos = mutations)

}

#Below line 'assumes' that mutatations in ORF1a AND ORF1AB, in the same position, with same peptide, are the same mutation, thus are joined to be ORF1AB/ORF1a
MUTATIONS=MUTATIONS %>% group_by(Peptide,VariantAlignment,start_pos,MutationPos) %>% dplyr::summarise(Protein = sort(paste0(Protein,collapse = "/")))
# Read in known SNPs to validate our findings. NAs are retained as they remain observed mutants. Future versions may seek to understand the NAs in more detail
MUTATION_LIST= fread("OMICRON_SNPS.csv")
MUTATION_LIST=MUTATION_LIST %>% mutate(MutationPos = parse_number(Mutation))
# Clean the data
MUTATIONS=MUTATIONS %>% separate_rows_(cols = "MutationPos",sep=",")%>%
  mutate(MutationPos =as.numeric(MutationPos)) %>%
  left_join(MUTATION_LIST)

MUTATIONS=MUTATIONS %>% group_by(Peptide, VariantAlignment,start_pos,Protein)%>%
  dplyr::summarise(MutationPos=paste0(MutationPos,collapse = ","),
                   Mutation = paste0(Mutation, collapse = ","))%>% ungroup

```

```{r}

MUTATIONS=MUTATIONS %>% mutate(Hamming = stringdist::stringdist(Peptide,VariantAlignment,method="hamming"))

#saveRDS(MUTATIONS,"OMICRON_EPITOPE_MUTATIONS.rds")

```

# Mutation statoistics

```{r,dpi=300}
# How many unique peptides with mutations
MUTATIONS %>% nrow

# Summarise the pep list, so group same positions of orf1a/orf1ab
SUMMARISED_WUHAN_PEPLIST=WUHAN_COV2_Ep_Coverage %>% select(Peptide,start_pos,end_pos,Protein,names)%>% group_by(Peptide,start_pos,end_pos) %>% dplyr::summarise(Protein = sort(paste0(Protein, collapse = "/")))%>% ungroup
#check no idiosyncracies. Looks good
SUMMARISED_WUHAN_PEPLIST %>% select(Protein) %>% unique
# should be 1380 obs, which there is
SUMMARISED_WUHAN_PEPLIST %>% nrow

```


```{r}

MissingPeps=setdiff(FullDataset$Peptide,WUHAN_COV2_Ep_Coverage$Peptide)

```

# Compile Antigen Presentation data table
## WT wuhan peptides

- DONE! - run stabpan and netmhcpan BA
- Now, Read in BA/STAB/COMBINED metrics for each peptide for the 50 alleles
### netMHCPAN
- Get BA 1-log50k and combined
-
```{r}

# Class I WUHAN

TEST_DATA_LOCATION = "SARS_COV2_DATA_DEC21/NETMHCPAN_CLASSI"
# Loop through all files (organised by HLA) and combine.
data_path <- TEST_DATA_LOCATION
files <- dir(data_path, pattern = "NetMHC_Results.csv")

data3 <- data_frame(file = files) %>%
  mutate(file_contents = map(file,
                             ~ fread(file.path(data_path, .),skip = 1))
  )

Netmhcpanres <- unnest(data3)
# Clean HLA info
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$file,pattern="Allele_|_NetMHC_Results.csv",replacement = ""))
# Change 1/0 binary binder/nonbinder to english
Netmhcpanres=Netmhcpanres %>% select(! c(file,Pos,ID,core,icore)) %>% mutate(Binder = ifelse(NB==1,"BINDER","NONBINDER"))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=HLA_Allele,pattern = "\\:",replacement = "")) # Clean HLA allele.

NETMHC_CLASSI=Netmhcpanres
# Determine a binder ourselves, by 2 threshold of BA Rank (instead of EL). Makes analysis of nM consistent with binding status.
NETMHC_CLASSI=NETMHC_CLASSI %>% mutate(Binder = ifelse(BA_Rank>=2, "NONBINDER","BINDER"))
# Calculate the nM from the BA score. BA score is 1-log(50k) of nM. Therefore, nM is 50k^ (1-BA score).
NETMHC_CLASSI=NETMHC_CLASSI %>% mutate(nM_41 = round(50000^(1-`BA-score`) ,digits=5))

# Summarise data
NETMHC_CLASSI= NETMHC_CLASSI %>% group_by(Peptide) %>% dplyr::rename(Predicted_Binding=HLA_Allele)%>%
  select(!c(NB)) %>% summarise(Predicted_Binding = paste0(unique(Predicted_Binding),collapse = ","),
          `EL-score`=paste0(`EL-score`,collapse = ","),
          `EL_Rank`=paste0(`EL_Rank`,collapse = ","),
          `BA-score`=paste0(`BA-score`,collapse = ","),
          `BA_Rank`=paste0(`BA_Rank`,collapse = ","),
          `Ave`=paste0(`Ave`,collapse = ","),
          `Binder` = paste0(`Binder`,collapse = ","),
          nM_41 = paste0(`nM_41`, collapse = ","))

NETMHC_CLASSI %>% head

#saveRDS(NETMHC_CLASSI,file="WUHAN_SARSCOV2_50ALLELES_NETMHCPAN.rds")

```

### Stabpan
- Read in stability scores

```{r}

TEST_DATA_LOCATION = "SARS_COV2_DATA_DEC21/STABPAN"
# Loop through all files (organised by HLA) and combine.
data_path <- TEST_DATA_LOCATION
files <- dir(data_path, pattern = "NetMHC_Results.csv")

data3 <- data_frame(file = files) %>%
  mutate(file_contents = map(file,
                             ~ fread(file.path(data_path, .),skip = 1))
  )

Netmhcpanres <- unnest(data3)
Netmhcpanres %>% head

# Clean HLA info from file name
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$file,pattern="Allele_|_NetMHC_Results.csv",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$HLA_Allele,pattern="LENGTH_[0-9]*_",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=HLA_Allele,pattern = "\\:",replacement = ""))

STABPANRES=Netmhcpanres %>% select(!file)

STABPANRES=STABPANRES %>% group_by(Peptide) %>% dplyr::rename(Predicted_Binding=HLA_Allele)%>%
  select(!c(NB)) %>%
  summarise(Predicted_Binding = paste0(unique(Predicted_Binding),collapse = ","),
          `Thalf(h)`=paste0(`Thalf(h)`,collapse = ","))

```

### Netmhcpan 4.0 BA

- Not really used but compiled anyway
```{r}

# Loop through all files (organised by HLA) and combine.
TEST_DATA_LOCATION = "SARS_COV2_DATA_DEC21/NETMHCPAN_CLASSI_AFFINITY"

data_path <- TEST_DATA_LOCATION
files <- dir(data_path, pattern = "NetMHC_Results.csv")

data3 <- data_frame(file = files) %>%
  mutate(file_contents = map(file,
                             ~ fread(file.path(data_path, .),skip = 1))
  )

Netmhcpanres <- unnest(data3)
Netmhcpanres %>% head
# Extract HLA allele from file name
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$file,pattern="Allele_|_NetMHC_Results.csv",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$HLA_Allele,pattern="LENGTH_[0-9]*_",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=HLA_Allele,pattern = "\\:",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(Aff_Binder = ifelse(NB==1,"BINDER","NONBINDER"))

AFFINITY=Netmhcpanres %>% select(Peptide, HLA_Allele, nM,Aff_Binder)

AFFINITY=AFFINITY %>% group_by(Peptide) %>% dplyr::rename(Predicted_Binding=HLA_Allele) %>% summarise(Predicted_Binding = paste0(unique(Predicted_Binding),collapse = ","),
          nM=paste0(nM,collapse = ","), Aff_Binder=paste0(Aff_Binder,collapse = ","))

```

### combine the binding scores

```{r}
STABPANRES %>% nrow
NETMHC_CLASSI%>% nrow
AFFINITY %>% nrow

STABPANRES %>% inner_join(NETMHC_CLASSI)  %>% nrow

STABPANRES %>% inner_join(NETMHC_CLASSI) %>% inner_join(AFFINITY)%>% nrow
WUHAN_PEPTIDES_BINDING_SCORES=STABPANRES %>% inner_join(NETMHC_CLASSI)%>% inner_join(AFFINITY)
WUHAN_PEPTIDES_BINDING_SCORES %>% head()

```

## compute fraction of hydrophobicity


```{r}

library(stringi)

HYDROPHOBIC_RESIDUES = c("V","I","L","F","M","W","C") # from BARNES 2003 (SEE WELLS PAPER)

WUHAN_PEPTIDES_BINDING_SCORES=WUHAN_PEPTIDES_BINDING_SCORES%>%
  mutate(Length = nchar(Peptide)) %>%
  mutate(HydrophobicCount =  stri_count_regex(WUHAN_PEPTIDES_BINDING_SCORES$Peptide, paste0(HYDROPHOBIC_RESIDUES,collapse = "|"))) %>%
  mutate(HydroFraction = HydrophobicCount/Length)

```

## Compile data

```{r}

WUHAN_PEPTIDES_BINDING_SCORES = WUHAN_PEPTIDES_BINDING_SCORES %>% filter(!Peptide %in% MissingPeps)
WUHAN_PEPTIDES_BINDING_SCORES %>% nrow
saveRDS(WUHAN_PEPTIDES_BINDING_SCORES, "ORIGINAL_PEPTIDES_BINDING_SCORES_V2.rds")
```


## Antigen presentation of MT peptides
- Repeat the above but for MT

- DONE! - run stabpan and netmhcpan BA
- Now, Read in BA/STAB/COMBINED metrics for each peptide for the 50 alleles

### netMHCPAN 4.1
- Get BA 1-log50k and combined
```{r}


# Class I OMICRON MUTANTS
TEST_DATA_LOCATION = "OUTPUT/OMICRON_NETMHC/NETMHCPAN_CLASSI"

data_path <- TEST_DATA_LOCATION
files <- dir(data_path, pattern = "NetMHC_Results.csv")

data3 <- data_frame(file = files) %>%
  mutate(file_contents = map(file,
                             ~ fread(file.path(data_path, .),skip = 1))
  )

Netmhcpanres <- unnest(data3)
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$file,pattern="Allele_|_NetMHC_Results.csv",replacement = ""))
Netmhcpanres=Netmhcpanres %>% select(! c(file,Pos,ID,core,icore)) %>% mutate(Binder = ifelse(NB==1,"BINDER","NONBINDER"))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=HLA_Allele,pattern = "\\:",replacement = ""))

NETMHC_CLASSI=Netmhcpanres
NETMHC_CLASSI=NETMHC_CLASSI %>% mutate(MT_nM_41 = round(50000^(1-`BA-score`) ,digits=5))
NETMHC_CLASSI=NETMHC_CLASSI %>% mutate(Binder = ifelse(BA_Rank>=2, "NONBINDER","BINDER"))

NETMHC_CLASSI %>% head
 NETMHC_CLASSI= NETMHC_CLASSI %>% dplyr::rename(VariantAlignment = Peptide)%>% group_by(VariantAlignment) %>%
         dplyr::rename(MT_Predicted_Binding=HLA_Allele,
                       `MT_EL-score`=`EL-score`,
                        `MT_EL_Rank`=`EL_Rank`,
                       `MT_BA-score`=`BA-score`,
                        `MT_BA_Rank`=`BA_Rank`,
                        `MT_Ave`=`Ave`,
                        `MT_Binder` = `Binder`) %>% select(!c(NB)) %>%
         summarise(MT_Predicted_Binding = paste0(unique(MT_Predicted_Binding),collapse = ","),
          `MT_EL-score`=paste0(`MT_EL-score`,collapse = ","),
          `MT_EL_Rank`=paste0(`MT_EL_Rank`,collapse = ","),
          `MT_BA-score`=paste0(`MT_BA-score`,collapse = ","),
          `MT_BA_Rank`=paste0(`MT_BA_Rank`,collapse = ","),
          `MT_Ave`=paste0(`MT_Ave`,collapse = ","),
          `MT_Binder` = paste0(`MT_Binder`,collapse = ","),
         MT_nM_41 = paste0(MT_nM_41, collapse = ","))

NETMHC_CLASSI %>% head
NETMHC_CLASSI %>% select(VariantAlignment) %>% distinct()%>% nrow
saveRDS(NETMHC_CLASSI,file="OMICRON_MUTANTS_SARSCOV2_50ALLELES_NETMHCPAN.rds")

```



### stabpan
- Read in stability scores

```{r}

TEST_DATA_LOCATION = "OUTPUT/OMICRON_NETMHC/STABPAN_MUTANT"

data_path <- TEST_DATA_LOCATION
files <- dir(data_path, pattern = "NetMHC_Results.csv")

data3 <- data_frame(file = files) %>%
  mutate(file_contents = map(file,
                             ~ fread(file.path(data_path, .),skip = 1))
  )

Netmhcpanres <- unnest(data3)
Netmhcpanres %>% head

Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$file,pattern="Allele_|_NetMHC_Results.csv",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$HLA_Allele,pattern="LENGTH_[0-9]*_",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=HLA_Allele,pattern = "\\:",replacement = ""))

Netmhcpanres %>% head
Netmhcpanres %>% tail
STABPANRES=Netmhcpanres %>% select(!file)
STABPANRES %>% head
STABPANRES%>% distinct() %>% group_by(Peptide) %>% dplyr::summarise(n=n_distinct(HLA_Allele))
STABPANRES=STABPANRES %>% dplyr::rename(VariantAlignment = Peptide, `MT_Thalf(h)`=`Thalf(h)`)%>% group_by(VariantAlignment)  %>% dplyr::rename(MT_Predicted_Binding=HLA_Allele)%>% select(!c(NB)) %>% summarise(MT_Predicted_Binding = paste0(unique(MT_Predicted_Binding),collapse = ","),
          `MT_Thalf(h)`=paste0(`MT_Thalf(h)`,collapse = ","))

```


### binding affinity from netMHCpan 4.0
- not used but was still compiled

```{r}


TEST_DATA_LOCATION = "OUTPUT/OMICRON_NETMHC/NETMHCPAN_CLASSI_AFFINITY"

data_path <- TEST_DATA_LOCATION
files <- dir(data_path, pattern = "NetMHC_Results.csv")

data3 <- data_frame(file = files) %>%
  mutate(file_contents = map(file,
                             ~ fread(file.path(data_path, .),skip = 1))
  )

Netmhcpanres <- unnest(data3)
Netmhcpanres %>% head

Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$file,pattern="Allele_|_NetMHC_Results.csv",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$HLA_Allele,pattern="LENGTH_[0-9]*_",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=HLA_Allele,pattern = "\\:",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(MT_Aff_Binder = ifelse(NB==1,"BINDER","NONBINDER"))

AFFINITY=Netmhcpanres %>% select(Peptide, HLA_Allele, nM,MT_Aff_Binder)

AFFINITY=AFFINITY %>% dplyr::rename(VariantAlignment = Peptide,MT_nM=nM)%>% group_by(VariantAlignment) %>% dplyr::rename(Predicted_Binding=HLA_Allele) %>% summarise(Predicted_Binding = paste0(unique(Predicted_Binding),collapse = ","),
          MT_nM=paste0(MT_nM,collapse = ","), MT_Aff_Binder=paste0(MT_Aff_Binder,collapse = ","))

```

```{r}
OMICRON_MUTATIONS_BINDING_SCORES = STABPANRES %>% inner_join(NETMHC_CLASSI) %>% inner_join(AFFINITY)
OMICRON_MUTATIONS_BINDING_SCORES %>% nrow
#OMICRON_MUTATIONS_BINDING_SCORES = OMICRON_MUTATIONS_BINDING_SCORES %>% filter(!Peptide %in% MissingPeps)


saveRDS(OMICRON_MUTATIONS_BINDING_SCORES, "OMICRON_PEPTIDES_BINDING_SCORES_V2.rds")

```

# 26 peptide discrepency with mapped to wuhan
- I have a 26 peptide discrepency between those in 'FullDataset' curated from IEDB+VIPR+MIRA, and those mapped to wuhan.
- Whats going on here?

- Either, misclassified from CoV1, from a variant e.g., lambda ("NYNYRYRLF"), etc, or non canonical
- These are thus excluded from further analysis.

```{r}


MissingPeps=setdiff(FullDataset$Peptide,WUHAN_COV2_Ep_Coverage$Peptide)
FullDataset=FullDataset %>% filter(!Peptide %in% MissingPeps)



```
















