---
title: "R Notebook"
author: paulbuckley
date: 14/12/2021
output: html_document
---

```{r}
library(data.table)
library(dplyr)
library(tidyr)
library(ggpubr)
library(Biostrings)
library(tidyverse)
library(foreach)
library(stringdist)
library(doParallel)

```
## Useful functions

```{r}

# Function to provide a closest match. Used to match HLA Alleles across mixed output styles.
ClosestMatch2 = function(string, stringVector){

  stringVector[amatch(string, stringVector, maxDist=Inf)]

}

```

## Aims
Question: - What SARS-CoV2 T cell targets are mutated in Omicron?

- Gather a comprehensive set of immunogenic SArs CoV2 peptides
- Compare Wuhan and Omicron
- What SARS-CoV2 T cell targets in mutated regions of Omicron?
-


# Gather data

- First, we need to gather a dataset of immunogenic CoV2 peptides
- Before this, I have combined VIPR + IEDB CoV2 human data as of 14/12/21, the PVE of which will be read in here
- We also incorporate any MIRA peptides.

## IEDB + VIPR

```{r}

IEDB_VIPR = readRDS("DATA/SARS_COV2_DATA_DEC21/DEC_2021_Human_Cleaned_SarsCoV2_EpitopeDataset_w_Freq_ONLYPOSITIVE.rds")
IEDB_VIPR_Full = IEDB_VIPR
IEDB_VIPR %>% DT::datatable()

IEDB_VIPR=IEDB_VIPR %>% select(Peptide) %>% mutate(Dataset = "IEDB+VIPR")

```

## MIRA
- Only take those not observed in IEDB+VIPR
-asumd immunogenic as got recognising TCRs
```{r}

MIRA_DATA_FULL =fread(file="DATA/MIRA_DATA_FULL.txt")
MIRA_MISSING=MIRA_DATA_FULL %>% select(Peptide)%>% distinct() %>% mutate(Dataset ="MIRA") %>% filter(! Peptide %in% IEDB_VIPR$Peptide)

MIRA_MISSING %>% head()
IEDB_VIPR_Full%>% head
```

## Join MIRA + IEDB+VIPR

```{r}
MIRA_MISSING_Full=MIRA_MISSING%>% mutate(HLA_Allele="Unknown", Protein_Name="Unknown", Qualitative_Measure = "Positive", Measurement = "MIRA",Immunogenicity_Binary="Positive",n=NA,Frequency_of_ImmunogenicityBinary=NA)
IEDB_VIPR_Full=IEDB_VIPR_Full %>% mutate(Dataset = "IEDB+VIPR")

MIRA_MISSING_Full %>% head
IEDB_VIPR_Full %>% head
IEDB_VIPR_MIRA_FULL=IEDB_VIPR_Full %>% mutate(Dataset = "IEDB+VIPR") %>% rbind(MIRA_MISSING_Full)
saveRDS(IEDB_VIPR_MIRA_FULL, file="IEDB_VIPR_MIRA_DATA_FULL_HLA.rds")
FullDataset = rbind(IEDB_VIPR,MIRA_MISSING)

saveRDS(FullDataset, "IEDB_VIPR_MIRA_COV2_DATASET.rds")

# filter only class I by length <15
FullDataset=FullDataset %>% mutate(Length = nchar(Peptide)) %>% filter(Length < 15)

FullDataset %>% filter(Peptide %in% "FLYIIKLVFL")


```



# Map epitopes to Wuhan isolate proteome

```{r}

full_peptides=FullDataset %>% select(Peptide) %>% unique %>% pull

#pull in CoV2 proteome (Wuhan)

CoV2_stringset=readAAStringSet("DATA/Coronavirus_Sequences/SARS-CoV-2-Wuhan.txt",use.names = TRUE)

Ep_Coverage = foreach(i = 1:length(full_peptides),.combine = rbind, .packages = c("dplyr","magrittr","data.table","Biostrings"))%do% {
  peptide_i=AAString(full_peptides[i])

  DT=as.data.table(unlist(vmatchPattern(peptide_i,CoV2_stringset)))%>% mutate(Peptide = toString(peptide_i))
  DT

}


```

```{r}

Ep_Coverage=Ep_Coverage %>% mutate(Protein = gsub(".*protein=(.+)\\].*","\\1",names))%>% mutate(Protein = gsub("\\].*","",Protein))
Ep_Coverage=Ep_Coverage %>% dplyr::rename(start_pos="start",end_pos="end")

Protein_positions=Ep_Coverage %>% group_by(r=row_number()) %>% mutate(allpos = list(start_pos:end_pos)) %>% unnest() %>% group_by(start_pos,end_pos) %>% dplyr::summarise(Positions = paste0(allpos,collapse = ","))%>% ungroup
WUHAN_COV2_Ep_Coverage=Ep_Coverage %>% inner_join(Protein_positions)

PROT_IDS=names(CoV2_stringset) %>% as.data.table() %>% dplyr::rename(names=".") %>% mutate(RefSeqProteinID = row_number())%>% mutate(Protein = gsub(".*protein=(.+)\\].*","\\1",names))%>% mutate(Protein = gsub("\\].*","",Protein))

WUHAN_COV2_Ep_Coverage=WUHAN_COV2_Ep_Coverage %>% inner_join(PROT_IDS)

```

## Cov2 peps: how many?
- by filtering length<15, we are left with 1380 unique peptides
- mapping to 1798 observations , given orf1a and orf1ab overlap

```{r}
WUHAN_COV2_Ep_Coverage%>% select(Peptide) %>% distinct()%>% nrow

WUHAN_COV2_Ep_Coverage %>% select(Peptide, Protein) %>% distinct() %>% nrow
#saveRDS(WUHAN_COV2_Ep_Coverage,file="WUHAN_COV2_Ep_Coverage.rds")

```



## CoV2 peps mapped to Wuhan, how many and where

- Highest num spike, highest num normalised by protein length is to M,N,S.

```{r,dpi=300}

WUHAN_COV2_Ep_Coverage %>% select(Peptide) %>% distinct %>% mutate(Length = nchar(Peptide)) %>% group_by(Length) %>% dplyr::summarise(n=n())%>%arrange(desc(n))%>% mutate(Length = as.character(Length))%>%
  ggbarplot(x="Length",y="n",fill="darkgrey")+ylab("Num. SARS-CoV-2 Peptides\nto Wuhan Isolate")+theme_pubr(base_size = 18)+coord_flip()

WUHAN_COV2_Ep_Coverage %>% select(Peptide, Protein) %>% distinct() %>% group_by(Protein) %>% dplyr::summarise(n=n())%>% arrange(desc(n)) %>%
  ggbarplot(x="Protein",y="n",fill="darkgrey")+ylab("Num. SARS-CoV-2 Peptides\nto Wuhan Isolate")+theme_pubr(base_size = 18)+coord_flip()

CORONAVIRUSSEQS = fread("DATA/Coronavirus_sequences_Full_DT.txt")%>% filter(Virus %in% 'SARS-CoV-2-Wuhan')
CORONAVIRUSSEQS=CORONAVIRUSSEQS %>% mutate(Length = nchar(`Seq.V1`)) %>% select(!c(Virus, `Seq.V1`))

WUHAN_COV2_Ep_Coverage %>% select(Peptide, Protein) %>% distinct() %>% group_by(Protein) %>% dplyr::summarise(n=n())%>% inner_join(CORONAVIRUSSEQS) %>% mutate(FreqPerAA = n/Length)%>% arrange(desc(FreqPerAA))%>%
   ggbarplot(x="Protein",y="FreqPerAA",fill="darkgrey")+ylab("Freq SARS-CoV-2 Peptides\nper Length Protein to Wuhan isolate")+theme_pubr(base_size = 18)+coord_flip()


```


# Map epitopes to the above regions of Omicron

## Import variant (Omicron) proteome
-Deal w/ proteinIDs which may differ between RefSeq (Wuhan) and Variant (Omicron)

```{r}

Variant_StringSet=readAAStringSet("Omicron_ProteinSeqs/Omicron_SARSCoV2_Seq.fasta",use.names = TRUE)
Variant_ProteinIDs = names(Variant_StringSet) %>% as.data.table() %>% dplyr::rename(Variantnames=".") %>% mutate(VariantProteinID = row_number())%>% mutate(Protein = gsub(".*protein=(.+)\\].*","\\1",Variantnames))%>% mutate(Protein = gsub("\\].*","",Protein))

```

```{r}

WUHAN_COV2_Ep_Coverage=WUHAN_COV2_Ep_Coverage %>% inner_join(Variant_ProteinIDs)

```


## local global to entire Variant

```{r}

VARIANT_PROTEINS = WUHAN_COV2_Ep_Coverage %>% select(Protein, VariantProteinID) %>% distinct()

cores=detectCores()
cl <- makeCluster(cores[1]-1) #not to overload your computer
registerDoParallel(cl)

Mutations_Epitopes = foreach(i = 1:nrow(WUHAN_COV2_Ep_Coverage),.combine = rbind, .packages = c("dplyr","magrittr","data.table","Biostrings")) %:%
foreach(ii = 1:length(Variant_StringSet),.combine = rbind, .packages = c("dplyr","magrittr","data.table","Biostrings"))%dopar% {
    peptide_i=WUHAN_COV2_Ep_Coverage$Peptide[i]
  AlignmentDT = NULL
  ALIGNMENT=pairwiseAlignment(Variant_StringSet[[ii]], peptide_i, type="local-global")
  start_match = start(pattern(ALIGNMENT))


  FullAlignmentDT = data.table(Peptide = peptide_i, VariantAlignment =   as.character(aligned(pattern(ALIGNMENT))), score=score(ALIGNMENT), TestedProtein=VARIANT_PROTEINS %>% filter(VariantProteinID == ii) %>% pull(Protein),VariantProteinID=VARIANT_PROTEINS %>% filter(VariantProteinID == ii) %>% pull(VariantProteinID), start_match = start_match )

}

stopCluster(cl)
Mutations_Epitopes = Mutations_Epitopes%>% distinct() %>% group_by(Peptide) %>% slice_max(score, with_ties = TRUE)%>% ungroup

Mutations_Epitopes=Mutations_Epitopes %>% dplyr::rename(Protein=TestedProtein) %>% select(Peptide, VariantAlignment, VariantProteinID, Protein,start_match)
Mutations_Epitopes = Mutations_Epitopes %>% distinct() %>% ungroup

```


## Mutation statistics


```{r}

Mutations_Epitopes %>% select(Peptide) %>% distinct()%>% nrow
Mutations_Epitopes %>% select(Peptide,Protein) %>% distinct()%>% nrow


Mutations_Epitopes=Mutations_Epitopes %>% mutate(Hamming = stringdist::stringdist(Peptide, VariantAlignment, method="hamming"))

Mutations_Epitopes %>% filter(!Hamming == 0)%>% select(Hamming) %>% nrow

Mutations_Epitopes %>% select(Hamming) %>% table
Mutations_Epitopes %>% select(Hamming) %>% table %>% prop.table()

# Join with data. Joins by proitein so any matches ! to the sdame protein are ignored. highly unlikely to be accurate matches.
MutationData_Wuhan_vs_Variant=WUHAN_COV2_Ep_Coverage %>% inner_join(Mutations_Epitopes)
MutationData_Wuhan_vs_Variant %>% nrow
MutationData_Wuhan_vs_Variant %>% select(Protein, Hamming) %>% table

MutationData_Wuhan_vs_Variant %>% select(Peptide) %>% distinct()%>% nrow
MutationData_Wuhan_vs_Variant %>% select(Peptide,Protein) %>% distinct()%>% nrow


#Mutations_Epitopes %>% filter(!Hamming == 0) %>% anti_join(WUHAN_COV2_Ep_Coverage) %>% nrow

#Mutations_Epitopes %>% filter(!Hamming == 0) %>% left_join(WUHAN_COV2_Ep_Coverage, by = c("Peptide","Protein")) %>% nrow


```



```{r}

MUTATIONDATA=MutationData_Wuhan_vs_Variant %>% filter(! Hamming==0)
MUTATIONDATA=MUTATIONDATA %>% mutate(ProportionMutated = Hamming/nchar(Peptide))
#MUTATIONDATA=MUTATIONDATA %>% filter(! ProportionMutated > 0.6)
calculateMutation <- function(str1, str2,start_pos) {
  str1vec <- unlist(strsplit(str1, ""))
  str2vec <- unlist(strsplit(str2, ""))
  iMut <- (1:length(str1vec))[str1vec != str2vec]
  MUTPOS = (start_pos-1)+iMut
  return(paste0(str1vec[iMut], MUTPOS, str2vec[iMut]))
}

MUTATIONDATA

calculateMutation(MUTATIONDATA$Peptide[1],MUTATIONDATA$VariantAlignment[1],653)

calculateMutation("HVSGTNGTK","HVISGTNGTK",69)



calculateMutation_position_V2 <- function(str1, str2,start_pos) {
  require(Biostrings)
  peptide = str1
  variantalignment = str2
  alignment = pairwiseAlignment(variantalignment,peptide)
  str1=as.character(aligned(pattern(alignment)))
  str2=as.character(aligned(subject(alignment)))



  str1vec <- unlist(strsplit(str1, ""))
  str2vec <- unlist(strsplit(str2, ""))
  iMut <- (1:length(str1vec))[str1vec != str2vec]
  MUTPOS = (start_pos-1)+iMut
  return(paste0(MUTPOS))
}

calculateMutation_position_V2("SSRGTSPAR","SSKRTSPAR",201)



```






```{r}

MUTATIONS = foreach(i = 1:nrow(MUTATIONDATA),.combine = rbind, .packages = c("dplyr","magrittr","data.table","Biostrings"))%do% {

  mutations = paste0(calculateMutation_position_V2(MUTATIONDATA$Peptide[i],MUTATIONDATA$VariantAlignment[i],MUTATIONDATA$start_pos[i]),collapse = ",")
  data.table(Peptide = MUTATIONDATA$Peptide[i], VariantAlignment=MUTATIONDATA$VariantAlignment[i], start_pos= MUTATIONDATA$start_pos[i], Protein = MUTATIONDATA$Protein[i], MutationPos = mutations)

}

#Below line assumes that mutatations in ORF1a and ORF1AB in the same position with same peptuide, are the same mutation thus are joined to be ORF1AB/ORF1a
MUTATIONS=MUTATIONS %>% group_by(Peptide,VariantAlignment,start_pos,MutationPos) %>% dplyr::summarise(Protein = sort(paste0(Protein,collapse = "/")))

MUTATION_LIST= fread("OMICRON_SNPS.csv")
MUTATION_LIST=MUTATION_LIST %>% mutate(MutationPos = parse_number(Mutation))
MUTATIONS %>% separate_rows_(cols = "MutationPos",sep=",")%>% mutate(MutationPos =as.numeric(MutationPos)) %>% nrow
MUTATIONS %>% filter(VariantAlignment %in% grep("\\-",VariantAlignment,value = T))
MUTATIONS=MUTATIONS %>% separate_rows_(cols = "MutationPos",sep=",")%>% mutate(MutationPos =as.numeric(MutationPos)) %>% left_join(MUTATION_LIST)
MUTATIONS=MUTATIONS %>% group_by(Peptide, VariantAlignment,start_pos,Protein)%>%
  dplyr::summarise(MutationPos=paste0(MutationPos,collapse = ","),
                   Mutation = paste0(Mutation, collapse = ","))%>% ungroup
MUTATIONS %>% nrow

```

```{r}

MUTATIONS=MUTATIONS %>% mutate(Hamming = stringdist::stringdist(Peptide,VariantAlignment,method="hamming"))

#saveRDS(MUTATIONS,"OMICRON_EPITOPE_MUTATIONS.rds")

```

# Mutation statoistics

```{r,dpi=300}
# How many unique peptides with mutations
MUTATIONS %>% nrow

# Summarise the pep list, so group same positions of orf1a/orf1ab
SUMMARISED_WUHAN_PEPLIST=WUHAN_COV2_Ep_Coverage %>% select(Peptide,start_pos,end_pos,Protein,names)%>% group_by(Peptide,start_pos,end_pos) %>% dplyr::summarise(Protein = sort(paste0(Protein, collapse = "/")))%>% ungroup
#check no idiosyncracies. Looks good
SUMMARISED_WUHAN_PEPLIST %>% select(Protein) %>% unique
# should be 1380 obs, which there is
SUMMARISED_WUHAN_PEPLIST %>% nrow

# Simply, what proportion of our original set (the distinct ones) are mutated?
SUMMARISED_WUHAN_PEPLIST=SUMMARISED_WUHAN_PEPLIST %>% mutate(MUTATED = ifelse(Peptide %in% MUTATIONS$Peptide, "Mutated","Conserved"))
SUMMARISED_WUHAN_PEPLIST %>% group_by(MUTATED) %>% dplyr::summarise(n=n())%>% ggbarplot(x="MUTATED",y="n",fill="darkgrey",label=T)+ylab("Number of Peptides")
SUMMARISED_WUHAN_PEPLIST %>% group_by(MUTATED) %>% dplyr::summarise(n=n())%>% mutate(Freq=n/sum(n))%>% ggbarplot(x="MUTATED",y="Freq",fill="darkgrey",label=T)+ylab("Frequency of total peptides")

# What is the distribution of number of mutation? - below first calculates how many mutations a peptide exhibits (NMUT), then sums the number of times x mutations has been observed (n)
MUTATIONS%>%select(Peptide,MutationPos) %>% separate_rows_(cols = "MutationPos",sep=",")%>% group_by(Peptide)%>% dplyr::summarise(NMUT = n_distinct(MutationPos))%>%ungroup %>%
        select(NMUT)%>% group_by(NMUT) %>% dplyr::summarise(n=n())%>%mutate(Freq=n/sum(n))

MUTATIONS%>%select(Peptide,MutationPos) %>% separate_rows_(cols = "MutationPos",sep=",")%>% group_by(Peptide)%>% dplyr::summarise(NMUT = n_distinct(MutationPos))%>%ungroup %>%
        select(NMUT)%>% group_by(NMUT) %>% dplyr::summarise(n=n())%>%mutate(Freq=n/sum(n)) %>% ggbarplot(x="NMUT",y="n",fill="darkgrey")+ylab("Number of Peptides")+xlab("Number of Mutations")

MUTATIONS %>% select(Peptide,VariantAlignment,Protein,MutationPos) %>% distinct() %>% group_by(Protein)%>% dplyr::summarise(n=n())%>% mutate(Freq=n/sum(n))
MUTATIONS %>% select(Peptide,VariantAlignment,Protein,MutationPos) %>% distinct() %>% group_by(Protein)%>% dplyr::summarise(n=n())%>% mutate(Freq=n/sum(n))%>%arrange(desc(n))%>%
  ggbarplot(x="Protein",y="n",fill="darkgrey")+ylab("Number of Peptides")+coord_flip()
```





```{r}

MissingPeps=setdiff(FullDataset$Peptide,WUHAN_COV2_Ep_Coverage$Peptide)

```



# How many mutations would we expect by random?

- per spike manner first
- % of spike peps with mutation in Omicron

```{r,cache=TRUE}

NUM_SPIKE_MUT=MUTATIONS %>% filter(Protein == 'surface glycoprotein') %>% nrow

#Griffoni data only
#NUM_SPIKE_EPS=FullDataset %>% filter(Restriction == 'CD8')%>% filter(Antigen == 'S') %>% nrow

NUM_SPIKE_EPS=WUHAN_COV2_Ep_Coverage %>% filter(Protein == 'surface glycoprotein') %>% nrow

NUM_SPIKE_MUT/NUM_SPIKE_EPS

MUTATIONS %>% filter(Protein == 'surface glycoprotein') %>% mutate(Length = nchar(Peptide))%>% select(Length) %>% table
WUHAN_COV2_Ep_Coverage %>% filter(Protein == 'surface glycoprotein')  %>% mutate(Length = nchar(Peptide))%>% select(Length)


# griffoni
#FullDataset %>% filter(Antigen == 'S') %>% mutate(Length = nchar(Peptide))%>% select(Length) %>% table

# do 9mers first

PEP_LENGTH = 9
PROTEIN = 'surface glycoprotein'

NUM_SPIKE_MUT=MUTATIONS %>% filter(Protein == PROTEIN ) %>% mutate(Length = nchar(Peptide)) %>% filter(Length == PEP_LENGTH)%>% nrow
#NUM_SPIKE_EPS=FullDataset %>% filter(Restriction == 'CD8')%>% filter(Antigen == 'S')%>% filter(Len == PEP_LENGTH) %>% nrow

NUM_SPIKE_EPS=WUHAN_COV2_Ep_Coverage %>% filter(Protein == PROTEIN )%>% mutate(Length = nchar(Peptide)) %>% filter(Length == PEP_LENGTH)%>% nrow

NUM_SPIKE_MUT/NUM_SPIKE_EPS


```



# Analysis 2
## WT wuhan peptides

- DONE! - run stabpan and netmhcpan BA
- Now, Read in BA/STAB/COMBINED metrics for each peptide for the 50 alleles
### netMHCPAN
- Get BA 1-log50k and combined
```{r}


# Class I WUHAN
TEST_DATA_LOCATION = "SARS_COV2_DATA_DEC21/NETMHCPAN_CLASSI"

data_path <- TEST_DATA_LOCATION
files <- dir(data_path, pattern = "NetMHC_Results.csv")

data3 <- data_frame(file = files) %>%
  mutate(file_contents = map(file,
                             ~ fread(file.path(data_path, .),skip = 1))
  )

 Netmhcpanres <- unnest(data3)
 Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$file,pattern="Allele_|_NetMHC_Results.csv",replacement = ""))
 Netmhcpanres=Netmhcpanres %>% select(! c(file,Pos,ID,core,icore)) %>% mutate(Binder = ifelse(NB==1,"BINDER","NONBINDER"))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=HLA_Allele,pattern = "\\:",replacement = ""))

 NETMHC_CLASSI=Netmhcpanres
NETMHC_CLASSI=NETMHC_CLASSI %>% mutate(Binder = ifelse(BA_Rank>=2, "NONBINDER","BINDER"))

NETMHC_CLASSI=NETMHC_CLASSI %>% mutate(nM_41 = round(50000^(1-`BA-score`) ,digits=5))
NETMHC_CLASSI %>% head
 NETMHC_CLASSI= NETMHC_CLASSI %>% group_by(Peptide) %>% dplyr::rename(Predicted_Binding=HLA_Allele)%>% select(!c(NB)) %>% summarise(Predicted_Binding = paste0(unique(Predicted_Binding),collapse = ","),
          `EL-score`=paste0(`EL-score`,collapse = ","),
          `EL_Rank`=paste0(`EL_Rank`,collapse = ","),
          `BA-score`=paste0(`BA-score`,collapse = ","),
          `BA_Rank`=paste0(`BA_Rank`,collapse = ","),
          `Ave`=paste0(`Ave`,collapse = ","),
          `Binder` = paste0(`Binder`,collapse = ","),
          nM_41 = paste0(`nM_41`, collapse = ","))

NETMHC_CLASSI %>% head

#saveRDS(NETMHC_CLASSI,file="WUHAN_SARSCOV2_50ALLELES_NETMHCPAN.rds")

```

### stabpan
- Read in stability scores

```{r}

TEST_DATA_LOCATION = "SARS_COV2_DATA_DEC21/STABPAN"

data_path <- TEST_DATA_LOCATION
files <- dir(data_path, pattern = "NetMHC_Results.csv")

data3 <- data_frame(file = files) %>%
  mutate(file_contents = map(file,
                             ~ fread(file.path(data_path, .),skip = 1))
  )

 Netmhcpanres <- unnest(data3)
Netmhcpanres %>% head

 Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$file,pattern="Allele_|_NetMHC_Results.csv",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$HLA_Allele,pattern="LENGTH_[0-9]*_",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=HLA_Allele,pattern = "\\:",replacement = ""))

Netmhcpanres %>% head
Netmhcpanres %>% tail
STABPANRES=Netmhcpanres %>% select(!file)
STABPANRES %>% head

STABPANRES=STABPANRES %>% group_by(Peptide) %>% dplyr::rename(Predicted_Binding=HLA_Allele)%>% select(!c(NB)) %>% summarise(Predicted_Binding = paste0(unique(Predicted_Binding),collapse = ","),
          `Thalf(h)`=paste0(`Thalf(h)`,collapse = ","))

```

### binding affinity

```{r}


TEST_DATA_LOCATION = "SARS_COV2_DATA_DEC21/NETMHCPAN_CLASSI_AFFINITY"

data_path <- TEST_DATA_LOCATION
files <- dir(data_path, pattern = "NetMHC_Results.csv")

data3 <- data_frame(file = files) %>%
  mutate(file_contents = map(file,
                             ~ fread(file.path(data_path, .),skip = 1))
  )

 Netmhcpanres <- unnest(data3)
Netmhcpanres %>% head

 Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$file,pattern="Allele_|_NetMHC_Results.csv",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$HLA_Allele,pattern="LENGTH_[0-9]*_",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=HLA_Allele,pattern = "\\:",replacement = ""))
 Netmhcpanres=Netmhcpanres %>% mutate(Aff_Binder = ifelse(NB==1,"BINDER","NONBINDER"))
Netmhcpanres %>% head

AFFINITY=Netmhcpanres %>% select(Peptide, HLA_Allele, nM,Aff_Binder)

AFFINITY=AFFINITY %>% group_by(Peptide) %>% dplyr::rename(Predicted_Binding=HLA_Allele) %>% summarise(Predicted_Binding = paste0(unique(Predicted_Binding),collapse = ","),
          nM=paste0(nM,collapse = ","), Aff_Binder=paste0(Aff_Binder,collapse = ","))

AFFINITY %>% head
```

### combine the binding scores

```{r}
STABPANRES %>% nrow
NETMHC_CLASSI%>% nrow
AFFINITY %>% nrow

STABPANRES %>% inner_join(NETMHC_CLASSI)  %>% nrow

STABPANRES %>% inner_join(NETMHC_CLASSI) %>% inner_join(AFFINITY)%>% nrow
WUHAN_PEPTIDES_BINDING_SCORES=STABPANRES %>% inner_join(NETMHC_CLASSI)%>% inner_join(AFFINITY)
WUHAN_PEPTIDES_BINDING_SCORES %>% head()

```

### compute fraction of hydrophobicity


```{r}

library(stringi)

HYDROPHOBIC_RESIDUES = c("V","I","L","F","M","W","C") # from BARNES 2003 (SEE WELLS PAPER)

WUHAN_PEPTIDES_BINDING_SCORES=WUHAN_PEPTIDES_BINDING_SCORES%>% mutate(Length = nchar(Peptide)) %>% mutate(HydrophobicCount =  stri_count_regex(WUHAN_PEPTIDES_BINDING_SCORES$Peptide, paste0(HYDROPHOBIC_RESIDUES,collapse = "|"))) %>% mutate(HydroFraction = HydrophobicCount/Length)

```
### Check ive got all scores for all peptides


```{r}
FullDataset%>% select(Peptide,Length) %>% distinct() %>% filter(Length<15)%>% nrow
WUHAN_PEPTIDES_BINDING_SCORES %>% nrow

```





```{r}

WUHAN_PEPTIDES_BINDING_SCORES %>% head()
WUHAN_PEPTIDES_BINDING_SCORES = WUHAN_PEPTIDES_BINDING_SCORES %>% filter(!Peptide %in% MissingPeps)
WUHAN_PEPTIDES_BINDING_SCORES %>% nrow
saveRDS(WUHAN_PEPTIDES_BINDING_SCORES, "ORIGINAL_PEPTIDES_BINDING_SCORES_V2.rds")
```

## MT peptides


- DONE! - run stabpan and netmhcpan BA
- Now, Read in BA/STAB/COMBINED metrics for each peptide for the 50 alleles
### netMHCPAN
- Get BA 1-log50k and combined
```{r}


# Class I OMICRON MUTANTS
TEST_DATA_LOCATION = "OUTPUT/OMICRON_NETMHC/NETMHCPAN_CLASSI"

data_path <- TEST_DATA_LOCATION
files <- dir(data_path, pattern = "NetMHC_Results.csv")

data3 <- data_frame(file = files) %>%
  mutate(file_contents = map(file,
                             ~ fread(file.path(data_path, .),skip = 1))
  )

 Netmhcpanres <- unnest(data3)
 Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$file,pattern="Allele_|_NetMHC_Results.csv",replacement = ""))
 Netmhcpanres=Netmhcpanres %>% select(! c(file,Pos,ID,core,icore)) %>% mutate(Binder = ifelse(NB==1,"BINDER","NONBINDER"))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=HLA_Allele,pattern = "\\:",replacement = ""))

 NETMHC_CLASSI=Netmhcpanres
NETMHC_CLASSI=NETMHC_CLASSI %>% mutate(MT_nM_41 = round(50000^(1-`BA-score`) ,digits=5))
NETMHC_CLASSI=NETMHC_CLASSI %>% mutate(Binder = ifelse(BA_Rank>=2, "NONBINDER","BINDER"))

NETMHC_CLASSI %>% head
 NETMHC_CLASSI= NETMHC_CLASSI %>% dplyr::rename(VariantAlignment = Peptide)%>% group_by(VariantAlignment) %>%
         dplyr::rename(MT_Predicted_Binding=HLA_Allele,
                       `MT_EL-score`=`EL-score`,
                        `MT_EL_Rank`=`EL_Rank`,
                       `MT_BA-score`=`BA-score`,
                        `MT_BA_Rank`=`BA_Rank`,
                        `MT_Ave`=`Ave`,
                        `MT_Binder` = `Binder`) %>% select(!c(NB)) %>%
         summarise(MT_Predicted_Binding = paste0(unique(MT_Predicted_Binding),collapse = ","),
          `MT_EL-score`=paste0(`MT_EL-score`,collapse = ","),
          `MT_EL_Rank`=paste0(`MT_EL_Rank`,collapse = ","),
          `MT_BA-score`=paste0(`MT_BA-score`,collapse = ","),
          `MT_BA_Rank`=paste0(`MT_BA_Rank`,collapse = ","),
          `MT_Ave`=paste0(`MT_Ave`,collapse = ","),
          `MT_Binder` = paste0(`MT_Binder`,collapse = ","),
         MT_nM_41 = paste0(MT_nM_41, collapse = ","))

NETMHC_CLASSI %>% head
NETMHC_CLASSI %>% select(VariantAlignment) %>% distinct()%>% nrow
saveRDS(NETMHC_CLASSI,file="OMICRON_MUTANTS_SARSCOV2_50ALLELES_NETMHCPAN.rds")

```



### stabpan
- Read in stability scores

```{r}

TEST_DATA_LOCATION = "OUTPUT/OMICRON_NETMHC/STABPAN_MUTANT"

data_path <- TEST_DATA_LOCATION
files <- dir(data_path, pattern = "NetMHC_Results.csv")

data3 <- data_frame(file = files) %>%
  mutate(file_contents = map(file,
                             ~ fread(file.path(data_path, .),skip = 1))
  )

 Netmhcpanres <- unnest(data3)
Netmhcpanres %>% head

 Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$file,pattern="Allele_|_NetMHC_Results.csv",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$HLA_Allele,pattern="LENGTH_[0-9]*_",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=HLA_Allele,pattern = "\\:",replacement = ""))

Netmhcpanres %>% head
Netmhcpanres %>% tail
STABPANRES=Netmhcpanres %>% select(!file)
STABPANRES %>% head
STABPANRES%>% distinct() %>% group_by(Peptide) %>% dplyr::summarise(n=n_distinct(HLA_Allele))
STABPANRES=STABPANRES %>% dplyr::rename(VariantAlignment = Peptide, `MT_Thalf(h)`=`Thalf(h)`)%>% group_by(VariantAlignment)  %>% dplyr::rename(MT_Predicted_Binding=HLA_Allele)%>% select(!c(NB)) %>% summarise(MT_Predicted_Binding = paste0(unique(MT_Predicted_Binding),collapse = ","),
          `MT_Thalf(h)`=paste0(`MT_Thalf(h)`,collapse = ","))

```


### binding affinity

```{r}


TEST_DATA_LOCATION = "OUTPUT/OMICRON_NETMHC/NETMHCPAN_CLASSI_AFFINITY"

data_path <- TEST_DATA_LOCATION
files <- dir(data_path, pattern = "NetMHC_Results.csv")

data3 <- data_frame(file = files) %>%
  mutate(file_contents = map(file,
                             ~ fread(file.path(data_path, .),skip = 1))
  )

 Netmhcpanres <- unnest(data3)
Netmhcpanres %>% head

 Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$file,pattern="Allele_|_NetMHC_Results.csv",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=Netmhcpanres$HLA_Allele,pattern="LENGTH_[0-9]*_",replacement = ""))
Netmhcpanres=Netmhcpanres %>% mutate(HLA_Allele = gsub(x=HLA_Allele,pattern = "\\:",replacement = ""))
 Netmhcpanres=Netmhcpanres %>% mutate(MT_Aff_Binder = ifelse(NB==1,"BINDER","NONBINDER"))

Netmhcpanres %>% head

AFFINITY=Netmhcpanres %>% select(Peptide, HLA_Allele, nM,MT_Aff_Binder)

AFFINITY=AFFINITY %>% dplyr::rename(VariantAlignment = Peptide,MT_nM=nM)%>% group_by(VariantAlignment) %>% dplyr::rename(Predicted_Binding=HLA_Allele) %>% summarise(Predicted_Binding = paste0(unique(Predicted_Binding),collapse = ","),
          MT_nM=paste0(MT_nM,collapse = ","), MT_Aff_Binder=paste0(MT_Aff_Binder,collapse = ","))
AFFINITY %>% head
```

```{r}
length(unique(MUTATIONS$VariantAlignment))
STABPANRES %>% nrow
NETMHC_CLASSI%>% nrow
AFFINITY %>% nrow
STABPANRES %>% inner_join(NETMHC_CLASSI) %>% inner_join(AFFINITY)%>% nrow

OMICRON_MUTATIONS_BINDING_SCORES = STABPANRES %>% inner_join(NETMHC_CLASSI) %>% inner_join(AFFINITY)
OMICRON_MUTATIONS_BINDING_SCORES %>% nrow
#OMICRON_MUTATIONS_BINDING_SCORES = OMICRON_MUTATIONS_BINDING_SCORES %>% filter(!Peptide %in% MissingPeps)


saveRDS(OMICRON_MUTATIONS_BINDING_SCORES, "OMICRON_PEPTIDES_BINDING_SCORES_V2.rds")

```
# 26 peptide discrepency with mapped to wuhan
- I have a 26 peptide discrepency between those in 'FullDataset' curated from IEDB+VIPR+MIRA, and those mapped to wuhan.
- Whats going on here?

- Either, misclassified from CoV1, from a variant e.g., lambda ("NYNYRYRLF"), etc, or non canonical
- These are thus excluded from further analysis.

```{r}


MissingPeps=setdiff(FullDataset$Peptide,WUHAN_COV2_Ep_Coverage$Peptide)
FullDataset=FullDataset %>% filter(!Peptide %in% MissingPeps)



```
















