---
title: "R Notebook"
output: html_document
---

```{r setup, include=FALSE}
library(dplyr)
library(data.table)
library(tidyr)
library(janitor)
library(doParallel)
library(caret)
library(stringdist)
library(ggpubr)
library(pROC)
library(yardstick)
# Function to provide a closest match. Used to match HLA Alleles/V genes across mixed output styles.
ClosestMatch2 = function(string, stringVector){

  stringVector[amatch(string, stringVector, maxDist=Inf)]

}
```


# Read in SUBVAR datasets

```{r}
# Reads in .rds file generated in 3_*.rmd.
# This is a DT containing for eah assesed variant, its mutated epitopes and WT-MT antigen presentation metrics.

FULL_MUTATIONS_DT = readRDS("SUBVARIANTS_FULL_MUTATIONS_DT.rds")
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>%  # Clean up the dataset for analysis
  select(Peptide, VariantAlignment, Predicted_Binding, BA_Rank,`BA-score`, Binder, MT_Binder, MT_BA_Rank,`MT_BA-score`,Variant)%>%
  separate_rows_(cols = c("Predicted_Binding", "BA_Rank","BA-score", "Binder", "MT_Binder", "MT_BA_Rank","MT_BA-score"),sep=",") %>%
  mutate(BA_Rank = as.numeric(BA_Rank)) %>% mutate(MT_BA_Rank = as.numeric(MT_BA_Rank))%>%mutate(`BA-score` = as.numeric(`BA-score`)) %>%
  mutate(`MT_BA-score` = as.numeric(`MT_BA-score`))

# Incorporate delta
DELTA_FULL_MUTATIONS_DT = readRDS("DELTA_FULL_MUTATIONS_DT.rds")
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>%  # Clean up the dataset for analysis
  select(Peptide, VariantAlignment, Predicted_Binding, BA_Rank,`BA-score`, Binder, MT_Binder, MT_BA_Rank,`MT_BA-score`,Variant)%>%
  separate_rows_(cols = c("Predicted_Binding", "BA_Rank","BA-score", "Binder", "MT_Binder", "MT_BA_Rank","MT_BA-score"),sep=",") %>%
  mutate(BA_Rank = as.numeric(BA_Rank)) %>% mutate(MT_BA_Rank = as.numeric(MT_BA_Rank))%>%mutate(`BA-score` = as.numeric(`BA-score`)) %>%
  mutate(`MT_BA-score` = as.numeric(`MT_BA-score`))

# Define negative logistic function. This transforms a 'BA Rank' score into a 0-1 probability based on binding rank threshold of 2.0 for cuttoff.
#negative logistic function
neg_logit_function = function(x) {
  1/(1+exp(1.5*(x-2)))
}
# Generate WT and MT MHCScore
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate(MHCScore = neg_logit_function(BA_Rank))%>% mutate(MT_MHCScore = neg_logit_function(MT_BA_Rank))
# Exclude WT&MT nonbinders - irrelevant for our analysis.
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT%>% filter(! (Binder == "NONBINDER" & MT_Binder == "NONBINDER"))
FULL_MUTATIONS_DT %>% head()
# Do delta separately
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT%>% mutate(MHCScore = neg_logit_function(BA_Rank))%>% mutate(MT_MHCScore = neg_logit_function(MT_BA_Rank))
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT%>% filter(! (Binder == "NONBINDER" & MT_Binder == "NONBINDER"))
DELTA_FULL_MUTATIONS_DT %>% head()
```

# Locate prediction scores from TRAP for each epitope

```{r}
# Read in full dataset of TRAP scores.
FullDataset=fread(cmd = "unzip -p TRAPP/DATA_V3_SARS2_ONLY/pred_scores/prott5_xl_bfd_peptides_sars2_mut_test_prediction_withlabel_2.txt.zip")%>% distinct()
# Get rid of mutagenesis scores, these are used later.
FullDataset=FullDataset %>% filter(! Dataset %in% 'Mutagenesis')
# Read in dataset that was used to generate TRAP prediction. These data re then linked with 'FullDataset', to find the MHC for each particular prediction score.
INPUT_DATA_TRAPP=fread("MUTAGENESIS_SUBSET_OMI_WUHAN_HLA_RUN_CHL_DNN_V4_INC_SUBVAR_WTNB_MTBINDER_ALLCOV2_MUTAGENESIS.txt")%>%
  distinct()%>% mutate(BA_Rank = as.numeric(BA_Rank))%>% mutate(nlog2Rank = - log2(BA_Rank))
# Round nlog2Rank [used as input to TRAP. Here I use to join with prediction score for correct MHC/
INPUT_DATA_TRAPP=INPUT_DATA_TRAPP %>% mutate(nlog2Rank = round(nlog2Rank, digits=4))
FullDataset=FullDataset%>% mutate(nlog2Rank = round(nlog2Rank, digits=4))
FullDataset=FullDataset %>% inner_join(INPUT_DATA_TRAPP)

FullDataset %>% select(Dataset)%>% table
# Cleaning and stats
FullDataset=FullDataset%>%dplyr::rename(Prediction=prediction)
FullDataset=FullDataset %>% select(Peptide, Dataset, Predicted_Binding, Prediction)
FullDataset %>% head
FullDataset=FullDataset %>% select(! Dataset) %>% distinct()

```

# Delta
- Read in TRAP predictions for delta
- Deal with delta separately
```{r}
delta_FullDataset=fread("TRAPP/DATA_V3_SARS2_ONLY/pred_scores/prott5_xl_bfd_peptides_sars2_delta_test_prediction.txt")%>% select(Peptide, Predicted_Binding, prediction)%>%
        dplyr::rename(Prediction=prediction)
```

# Generate immunogenicity prediction scores for all mutated epitopes
- Join FullMutationsDT with our scores
- Then where necessary, impute a immunogenicity score
- Below is for BA2, 4, 5
```{r}
# Round all numerics. Avoids duplicate observations
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate_if(is.numeric, round, 4)
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate(Length = nchar(Peptide)) %>% filter(Length %in% c(9,10))

FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% left_join(FullDataset)

# Simply impute
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% left_join(FullDataset %>% dplyr::rename(VariantAlignment = Peptide, OmicronPrediction=Prediction))
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate_if(is.numeric, round, 4)
# If a TRAP score does not exist, i.e is NA, then impute 0.01.
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% dplyr::rename(TRAPP_Prediction=Prediction, TRAPP_Omicron = OmicronPrediction) %>% replace_na(list(TRAPP_Prediction = 0.01, TRAPP_Omicron=0.01))
# Round again
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate_if(is.numeric, round, 4)
# Calculate immunogenicity score: TRAP * MHC
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate(Prediction = TRAPP_Prediction * MHCScore)%>% mutate(OmicronPrediction = TRAPP_Omicron* MT_MHCScore)
# Round again
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate_if(is.numeric, round, 4)
# Now, to avoid strange data, for prediction scores lower than 1e-05, replace with 1e-05. This is to avoid some scores with e.g., 1e-16.
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate(Prediction = replace(Prediction, Prediction<0.01, 0.01))%>%
  mutate(OmicronPrediction = replace(OmicronPrediction, OmicronPrediction<0.01, 0.01))
# Exclude any observations where immunogenicity score of WT and MT are both <0.5 - suggests nonimmungoenic binders.. but we're only interested in case where one is predicted immunogenic.
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% filter(!(Prediction < 0.5 & OmicronPrediction < 0.5))
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate(Prediction = round(Prediction, digits=4))%>% mutate(OmicronPrediction = round(OmicronPrediction, digits=4))%>% distinct()
# Anything rounded to 0.0, replace with 1e-05. To avoid issues with plotting
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate(Prediction = replace(Prediction, Prediction==0.0, 0.01))%>%
  mutate(OmicronPrediction = replace(OmicronPrediction, OmicronPrediction==0.0, 0.01))
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate_if(is.numeric, round, 4)
FULL_MUTATIONS_DT %>% dplyr::rename("Variant Peptide"=VariantAlignment, "HLA Allele"=Predicted_Binding,WuhanPrediction=Prediction,TRAPP_Wuhan=TRAPP_Prediction) %>%
    select(! c(`BA-score`,`MT_BA-score`))%>%
    readr::write_csv(file="/Users/paulbuckley/Nexus365/WIMM CCB - Koohy Group - Documents/Koohy Group/Effect_Mutation_Tcells/V11/SUBMITTED/R_R/Supplementary Datafile/Table3_Subvar_Tcellimmuno.csv")

```

## Plot Wuhan -->  BA2/4/5 changes in immunogenicity

```{r,dpi=300, fig.width = 12}
my_comparisons = list(c("Wuhan","Omicron"))
# For each variant, violin plot comparing immunogenicity score of WT vs. MT. (Wuhan vs Omicron)
VIOLIN_PLT=FULL_MUTATIONS_DT %>% select(Peptide, Variant, Predicted_Binding, Prediction, OmicronPrediction)%>%
        dplyr::rename(Wuhan=Prediction, Omicron=OmicronPrediction)%>%
        melt%>% dplyr::rename(Dataset=variable, Prediction=value)%>%
  ggviolin(x="Dataset",y="Prediction",add="boxplot",trim=TRUE)+
        stat_compare_means(comparisons = my_comparisons,label="p.signif")+
        geom_hline(yintercept = 0.75, linetype="dashed")+facet_grid(~Variant)+ylim(0,1)
VIOLIN_PLT

#saveRDS(VIOLIN_PLT,file="SUBVAR_WUHAN_OMI_VIOLIN_PLT.rds")
# For each variant, EDDF comparing immunogenicity score of WT vs. MT. (Wuhan vs Omicron)
ECDF_PLT=FULL_MUTATIONS_DT  %>% select(Peptide, Variant, Predicted_Binding, Prediction, OmicronPrediction)%>%dplyr::rename(Wuhan=Prediction, Omicron=OmicronPrediction)%>%
        melt%>% dplyr::rename(Dataset=variable, Prediction=value) %>%
        ggplot(aes(x=Prediction, color=Dataset))+stat_ecdf(size=1)+theme_pubr(base_size = 16)+grids()+geom_vline(xintercept = 0.75,linetype="dashed")+facet_grid(~Variant)
ECDF_PLT
#saveRDS(ECDF_PLT,file="SUBVAR_ECDF_PLT_IMMUNO.rds")

```


# Analyse HLA associations given changes in immunogenicity

```{r,dpi=300,fig.height=11,fig.width=10}
#Link the data with HLA supertypes
AB_SUPERTYPES=fread("HLA_AB_SUPERTYPES.csv") %>% mutate(Allele = gsub("\\*","",Allele))%>% mutate(Allele = paste0("HLA-",Allele))%>% dplyr::rename(HLA_Allele = Allele)

DATA_FOR_ANALYSIS=FULL_MUTATIONS_DT

HLA_A_B_AGRETOPICITY=DATA_FOR_ANALYSIS %>% dplyr::rename(HLA_Allele=Predicted_Binding)%>% filter(HLA_Allele %in% grep("HLA-A|HLA-B",HLA_Allele,value = T)) %>% inner_join(AB_SUPERTYPES)

HLA_CDATASET_AGRETOPICITY=DATA_FOR_ANALYSIS %>% dplyr::rename(HLA_Allele=Predicted_Binding)%>% filter(!HLA_Allele %in% grep("HLA-A|HLA-B",HLA_Allele,value = T))%>% mutate(Supertype = stringr::str_extract(HLA_Allele,"HLA-C[0-9][0-9]"))%>%
        mutate(Supertype = gsub("HLA\\-","",Supertype))

```

```{r,dpi=300, fig.width  = 11, fig.height = 11}
# For a set of supertypes, plot paired responses between
SUBVAR_SUPERTYPE_BOXPLT=HLA_A_B_AGRETOPICITY %>% rbind(HLA_CDATASET_AGRETOPICITY)%>%
        dplyr::rename(Wuhan=Prediction, Omicron=OmicronPrediction)%>% filter(Supertype %in% c("A02","A03","B07","C01"))%>%
        mutate(LRabs=abs(log(Omicron/Wuhan)))%>%mutate(label = ifelse(LRabs>4.0, Peptide, ""))%>%
        ggpaired(cond1 = "Wuhan", cond2 = "Omicron",line.color = "gray")+facet_wrap(Variant~Supertype)+
        theme_pubr(base_size = 16)+stat_compare_means(label = "p.signif",label.x.npc = "center")+
        rotate_x_text(angle=45)+ylab("Immunogenicity Score")

SUBVAR_SUPERTYPE_BOXPLT
#saveRDS(SUBVAR_SUPERTYPE_BOXPLT,file="SUBVAR_SUPERTYPE_BOXPLT_TCELL.rds")
SUBVAR_SUPERTYPE_BOXPLT=HLA_A_B_AGRETOPICITY %>% rbind(HLA_CDATASET_AGRETOPICITY)%>%
        dplyr::rename(Wuhan=Prediction, Omicron=OmicronPrediction)%>% filter(Supertype %in% c("A02","A03","B07","C01"))%>%
        mutate(LRabs=abs(log(Omicron/Wuhan)))%>%mutate(label = ifelse(LRabs>4.0, Peptide, ""))%>%mutate(label = substr(label, 1, 4))%>%
        ggpaired(cond1 = "Wuhan", cond2 = "Omicron",line.color = "gray")+facet_wrap(Variant~Supertype)+
        theme_pubr(base_size = 16)+stat_compare_means(label = "p.signif",label.x.npc = "center")+
        rotate_x_text(angle=45)+ylab("Immunogenicity Score")+ggrepel::geom_text_repel(max.overlaps = 6,aes(label=label))

SUBVAR_SUPERTYPE_BOXPLT

```



```{r,dpi=300, fig.width = 10, fig.height = 10}
MUTATIONS = readRDS("OMICRON_EPITOPE_MUTATIONS.rds")
BA1_OMICRON_MUT_LIST = MUTATIONS%>% separate_rows(Mutation, sep=",") %>% pull(Mutation) %>% unique

SUBVAR_MUTATIONS = readRDS("OMICRON_EPITOPE_MUTATIONS_SUBVAR.rds")

```

# Look for associations with immunogenicity changes by mutation

```{r,dpi=300, fig.height = 15, fig.width = 24}
# Identify mutations with sig differences Wuhan vs Omicron subvariants
MUTS_SIG_FOR_PLT = FULL_MUTATIONS_DT %>% inner_join(SUBVAR_MUTATIONS)%>% separate_rows(Mutation, sep=",") %>%
        filter(! Mutation %in% BA1_OMICRON_MUT_LIST)%>% dplyr::rename(Wuhan=Prediction,Omicron=OmicronPrediction) %>%
        pivot_longer(cols = c("Wuhan","Omicron"))%>%dplyr::rename(Dataset=name, Prediction=value)%>%
        separate_rows(Mutation, sep=",")%>%
        group_by(Mutation)%>%
        rstatix::wilcox_test(Prediction ~ Dataset, paired=TRUE) %>% filter(p<0.05) %>% pull(Mutation)
# Plot these changes
SUBVAR_MUTATIONS_COMPARE_PLT=FULL_MUTATIONS_DT %>% inner_join(SUBVAR_MUTATIONS)%>% separate_rows(Mutation, sep=",") %>%
        filter(! Mutation %in% BA1_OMICRON_MUT_LIST)%>% # only those distinct to a subvariant
        select(Peptide, VariantAlignment, Predicted_Binding, Mutation, Prediction, OmicronPrediction,Variant)%>%
        dplyr::rename(Wuhan=Prediction, Omicron=OmicronPrediction)%>%melt %>%dplyr::rename(Dataset = variable)%>%
        filter(Mutation %in% MUTS_SIG_FOR_PLT)%>%filter(! Mutation == "NA")%>%
 ggviolin(x="Dataset",y="value",fill="Dataset",add=c("jitter","boxplot"), trim=TRUE)+
        theme_pubr(base_size = 16)+facet_wrap(~Mutation+Variant,scales="free",ncol=6)+
        stat_compare_means(label="p.signif", label.x.npc = "center")+ylab("Immunogenicity Score")
SUBVAR_MUTATIONS_COMPARE_PLT

#saveRDS(SUBVAR_MUTATIONS_COMPARE_PLT, file="SUBVAR_MUTATIONS_COMPARE_PLT.rds")
```

# Compute RI score
- Below is called "LR Immuno"

```{r}

# Round up for log ratios. Anything < 0.01, set to 0.01.
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate(Prediction = replace(Prediction, Prediction<0.01, 0.01))%>%
  mutate(OmicronPrediction = replace(OmicronPrediction, OmicronPrediction<0.01, 0.01))
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% filter(!(Prediction < 0.5 & OmicronPrediction < 0.5))
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate(Prediction = round(Prediction, digits=4))%>% mutate(OmicronPrediction = round(OmicronPrediction, digits=4))%>% distinct()
FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate(Prediction = replace(Prediction, Prediction==0.0, 0.01))%>%
  mutate(OmicronPrediction = replace(OmicronPrediction, OmicronPrediction==0.0, 0.01))

FULL_MUTATIONS_DT=FULL_MUTATIONS_DT %>% mutate_if(is.numeric, round, 4)%>% distinct()

FITNESS_DATA=FULL_MUTATIONS_DT %>% dplyr::rename(WuhanPrediction=Prediction)
FITNESS_DATA=FITNESS_DATA%>%
       mutate(LR_Immuno = log(OmicronPrediction / WuhanPrediction)) # This is RI Score
FITNESS_DATA=FITNESS_DATA%>% mutate(LR_Immuno = round(LR_Immuno,digits = 4))%>% distinct()

#saveRDS(FITNESS_DATA,file="LR_IMMUNO_FITNESS_DATA_TRAPP_SUBVARS.rds")
FITNESS_DATA %>% DT::datatable()


```

## Visualise RI score (LRImmuno)

```{r,dpi=300, fig.width = 24,fig.height=6}

SUPERTYPE=fread("HLA_ABC_SUPERTYPES_CLEANED.csv")%>% dplyr::rename(Predicted_Binding=HLA_Allele)
NEW_COL=RColorBrewer::brewer.pal(3, "Set1")
NEW_COL = c("#377EB8","#E41A1C","#4DAF4A")

VARIANT = "BA2"

LRIMMUNO_PLT= FITNESS_DATA%>% filter(Variant == VARIANT)%>% mutate(pMHC=paste0(Peptide,"_",gsub("HLA-","",Predicted_Binding)))%>%arrange(desc(LR_Immuno))%>% left_join(SUPERTYPE)%>% left_join(SUBVAR_MUTATIONS)%>%
        filter(Protein == 'surface glycoprotein')%>%ungroup()%>% arrange(desc(LR_Immuno))%>%drop_na()%>% distinct()%>%mutate(Binder_Grp = paste0("WT_",Binder," MT_",MT_Binder))%>%
        ggplot(aes(x=reorder(pMHC, -LR_Immuno), y=LR_Immuno, fill=Binder_Grp))+geom_bar(stat="identity")+theme_pubr()+font("x.text",size=10)+rotate_x_text(angle=90)+theme(legend.position = "top")+grids()+xlab("Peptide-MHC")+ scale_fill_manual(values = NEW_COL)+ scale_x_discrete(guide = guide_axis(check.overlap = TRUE))+ggtitle(VARIANT)
LRIMMUNO_PLT
#saveRDS(LRIMMUNO_PLT, file = paste0(VARIANT,"_LRImmuno_SUBVAR_PLT.rds"))
VARIANT = "BA4"
LRIMMUNO_PLT= FITNESS_DATA%>% filter(Variant == VARIANT)%>% mutate(pMHC=paste0(Peptide,"_",gsub("HLA-","",Predicted_Binding)))%>%arrange(desc(LR_Immuno))%>% left_join(SUPERTYPE)%>% left_join(SUBVAR_MUTATIONS)%>%
        filter(Protein == 'surface glycoprotein')%>%ungroup()%>% arrange(desc(LR_Immuno))%>%drop_na()%>% distinct()%>%mutate(Binder_Grp = paste0("WT_",Binder," MT_",MT_Binder))%>%
        ggplot(aes(x=reorder(pMHC, -LR_Immuno), y=LR_Immuno, fill=Binder_Grp))+geom_bar(stat="identity")+theme_pubr()+font("x.text",size=10)+rotate_x_text(angle=90)+theme(legend.position = "top")+grids()+xlab("Peptide-MHC")+ scale_fill_manual(values = NEW_COL)+ scale_x_discrete(guide = guide_axis(check.overlap = TRUE))+ggtitle(VARIANT)#+scale_x_discrete(guide = guide_axis(n.dodge=2))
LRIMMUNO_PLT
#saveRDS(LRIMMUNO_PLT, file = paste0(VARIANT,"_LRImmuno_SUBVAR_PLT.rds"))

VARIANT = "BA5"
LRIMMUNO_PLT= FITNESS_DATA%>% filter(Variant == VARIANT)%>% mutate(pMHC=paste0(Peptide,"_",gsub("HLA-","",Predicted_Binding)))%>%arrange(desc(LR_Immuno))%>% left_join(SUPERTYPE)%>% left_join(SUBVAR_MUTATIONS)%>%
        filter(Protein == 'surface glycoprotein')%>%ungroup()%>% arrange(desc(LR_Immuno))%>%drop_na()%>% distinct()%>%mutate(Binder_Grp = paste0("WT_",Binder," MT_",MT_Binder))%>%
        ggplot(aes(x=reorder(pMHC, -LR_Immuno), y=LR_Immuno, fill=Binder_Grp))+geom_bar(stat="identity")+theme_pubr()+font("x.text",size=10)+rotate_x_text(angle=90)+theme(legend.position = "top")+grids()+xlab("Peptide-MHC")+ scale_fill_manual(values = NEW_COL)+ scale_x_discrete(guide = guide_axis(check.overlap = TRUE))+ggtitle(VARIANT)
LRIMMUNO_PLT
#saveRDS(LRIMMUNO_PLT, file = paste0(VARIANT,"_LRImmuno_SUBVAR_PLT.rds"))

```

# Develop a 'PAN-PEPTIDE' LR_Immuno

- Above analyses LRImmuno in a pMHC-specific manner. However for some applications (e.g., TCR Repertoire analysis), we are unable to perfectly integrate TCR-peptide-pMHC-LRImuno score, with patient HLAs.
- Therefore, instead, we must develop a score for each peptide, rather than each pMHC.
- To do this, I will use the same variables, WTFit, MTFit and LR_Immuno. For now the only change i will make, is for each _MHC_BINIDING_PROB, I will take an average of each peptide's binding scores across all HLA, rather than a MHC-specific score.


```{r}

FULL_DT_WIDE = FULL_MUTATIONS_DT%>% dplyr::rename(WuhanPrediction=Prediction)
FULL_DT_WIDE %>% DT::datatable()
FITNESS_DATA=FULL_DT_WIDE  %>% group_by(Peptide, VariantAlignment,Variant)%>%dplyr::summarise(AvgWuhan=mean(WuhanPrediction), AvgOmicron=mean(OmicronPrediction))%>%
        ungroup

FITNESS_DATA=FITNESS_DATA%>%
        mutate(LR_Immuno = log(AvgOmicron / AvgWuhan))

FITNESS_DATA %>% DT::datatable()

#saveRDS(FITNESS_DATA,file="LR_IMMUNO_FITNESS_DATA_PAN_PEPTIDE_SUBVAR.rds")

```

# Analyse Delta changes in immunogenicity

```{r}
# Round all numerics. Avoids duplicate observations
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% mutate_if(is.numeric, round, 4)
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% mutate(Length = nchar(Peptide)) %>% filter(Length %in% c(9,10))

DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% left_join(delta_FullDataset)
DELTA_FULL_MUTATIONS_DT
# Simply impute
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% left_join(delta_FullDataset %>% dplyr::rename(VariantAlignment = Peptide, OmicronPrediction=Prediction))
DELTA_FULL_MUTATIONS_DT
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% mutate_if(is.numeric, round, 4)
# If a TRAP score does not exist, i.e is NA, then impute 0.01.
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% dplyr::rename(TRAPP_Prediction=Prediction, TRAPP_Omicron = OmicronPrediction) %>% replace_na(list(TRAPP_Prediction = 0.01, TRAPP_Omicron=0.01))
# Round again
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% mutate_if(is.numeric, round, 4)
# Calculate immunogenicity score: TRAP * MHC
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% mutate(Prediction = TRAPP_Prediction * MHCScore)%>% mutate(OmicronPrediction = TRAPP_Omicron* MT_MHCScore)
# Round again
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% mutate_if(is.numeric, round, 4)
# Now, to avoid strange data, for prediction scores lower than 1e-05, replace with 1e-05. This is to avoid some scores with e.g., 1e-16.
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% mutate(Prediction = replace(Prediction, Prediction<0.01, 0.01))%>%
  mutate(OmicronPrediction = replace(OmicronPrediction, OmicronPrediction<0.01, 0.01))
# Exclude any observations where immunogenicity score of WT and MT are both <0.5 - suggests nonimmungoenic binders.. but we're only interested in case where one is predicted immunogenic.
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% filter(!(Prediction < 0.5 & OmicronPrediction < 0.5))
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% mutate(Prediction = round(Prediction, digits=4))%>% mutate(OmicronPrediction = round(OmicronPrediction, digits=4))%>% distinct()
# Anything rounded to 0.0, replace with 1e-05. To avoid issues with plotting
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% mutate(Prediction = replace(Prediction, Prediction==0.0, 0.01))%>%
  mutate(OmicronPrediction = replace(OmicronPrediction, OmicronPrediction==0.0, 0.01))
DELTA_FULL_MUTATIONS_DT=DELTA_FULL_MUTATIONS_DT %>% mutate_if(is.numeric, round, 4)
DELTA_FULL_MUTATIONS_DT
```


```{r,dpi=300, fig.width = 12}
my_comparisons = list(c("Wuhan","Delta"))
# For each variant, violin plot comparing immunogenicity score of WT vs. MT. (Delta vs Omicron)
DELTAVIOLIN_PLT=DELTA_FULL_MUTATIONS_DT %>% select(Peptide, Variant, Predicted_Binding, Prediction, OmicronPrediction)%>%
        dplyr::rename(Wuhan=Prediction, Delta=OmicronPrediction)%>%
        melt%>% dplyr::rename(Dataset=variable, Prediction=value)%>%
  ggviolin(x="Dataset",y="Prediction",add="boxplot",trim=TRUE)+
        stat_compare_means(comparisons = my_comparisons,label="p.signif")+
        geom_hline(yintercept = 0.75, linetype="dashed")+facet_grid(~Variant)+ylim(0,1)
DELTAVIOLIN_PLT

# For each variant, EDDF comparing immunogenicity score of WT vs. MT. (Delta vs Omicron)
DELTAECDF_PLT=DELTA_FULL_MUTATIONS_DT  %>% select(Peptide, Variant, Predicted_Binding, Prediction, OmicronPrediction)%>%dplyr::rename(Wuhan=Prediction, Delta=OmicronPrediction)%>%
        melt%>% dplyr::rename(Dataset=variable, Prediction=value) %>%
        ggplot(aes(x=Prediction, color=Dataset))+stat_ecdf(size=1)+theme_pubr(base_size = 16)+grids()+geom_vline(xintercept = 0.75,linetype="dashed")+facet_grid(~Variant)+theme(legend.position = "right")
DELTAECDF_PLT
my_comparisons = list(c("Wuhan","Omicron"))

```

```{r,fig.width=20,dpi=300}

cowplot::plot_grid(DELTAVIOLIN_PLT, DELTAECDF_PLT, NULL,nrow=1, rel_widths = c(0.25,0.35,0.4), align="hv",axis="bl")

```












