---
title: "R Notebook"
output: html_document
---

```{r}
library(data.table)
library(dplyr)
library(tidyr)
library(ggpubr)
library(Biostrings)
library(tidyverse)
library(foreach)
library(stringdist)
library(doParallel)

```



# Read in Wuhan binding dataset
- Sample of what the data look like.
```{r}

WUHAN_PEPTIDES_BINDING_SCORES=readRDS("ORIGINAL_PEPTIDES_BINDING_SCORES_V2.rds")
WUHAN_PEPTIDES_BINDING_SCORES %>% head
WUHAN_PEPTIDES_BINDING_SCORES %>% nrow


```



# Read in Omicron mutanta binding dataset
- Sample of what the data look like.
```{r}

OMICRON_PEPTIDES_BINDING_SCORES=readRDS("OMICRON_PEPTIDES_BINDING_SCORES_V2.rds")
OMICRON_PEPTIDES_BINDING_SCORES %>% head
OMICRON_PEPTIDES_BINDING_SCORES %>% nrow


```






# Read in mutations

```{r}
MUTATIONS = readRDS("OMICRON_EPITOPE_MUTATIONS.rds")
MUTATIONS %>% mutate(Length = nchar(Peptide))%>% filter(Protein == 'surface glycoprotein') %>% filter(Length== 9)%>% nrow

```


# join all
- Link mutations and agretopiocityDT. Therefore generating 'FULL_MUTATIONS_DT"
- FULL_MUTATIONS_DT contains WT-MT paired obs, with AgPres metrics of the WT and MT. Mutant columns are prefixed with MT_
```{r}

MUTATIONS %>% inner_join(WUHAN_PEPTIDES_BINDING_SCORES)%>% nrow
MUTATIONS %>% inner_join(OMICRON_PEPTIDES_BINDING_SCORES) %>% nrow
# Join MUTATIONS and WT & MT Ag Pres
FULL_MUTATIONS_DT=MUTATIONS %>% inner_join(WUHAN_PEPTIDES_BINDING_SCORES) %>% inner_join(OMICRON_PEPTIDES_BINDING_SCORES)
# Create AgretopicityDT
AGRETOPICITY_DT=FULL_MUTATIONS_DT %>% select(Peptide, VariantAlignment, Protein, Predicted_Binding, nM_41, MT_nM_41,Binder,MT_Binder) %>%
        separate_rows_(cols = c("Predicted_Binding","nM_41","MT_nM_41","Binder","MT_Binder"),sep=",") %>% # one row per obs
        mutate(nM_41 = as.numeric(nM_41)) %>% mutate(MT_nM_41 = as.numeric(MT_nM_41))
AGRETOPICITY_DT=AGRETOPICITY_DT %>% mutate(Agretopicity = MT_nM_41/nM_41)
# Spike DT only
SPIKE_AGRETOPICITY=AGRETOPICITY_DT %>% filter(Protein == 'surface glycoprotein') %>% mutate(Length = nchar(Peptide))



```


# FIG2A

```{r,dpi=300, fig.width = 8,fig.height=6}

SPIKE_AGRETOPICITY=SPIKE_AGRETOPICITY %>% filter(! (Binder == "NONBINDER" & MT_Binder == "NONBINDER"))
# ECDF comparing BA1 Wuhan vs Mutated
ECDF_NM_PLT=SPIKE_AGRETOPICITY %>% select(Peptide, nM_41, MT_nM_41)%>%
  pivot_longer(cols = c(nM_41,MT_nM_41), names_to = "Dataset",values_to = "Aff")%>%
  mutate(Dataset = replace(Dataset, Dataset == "nM_41", "Wuhan"))%>% # Change col name for visualisation
  mutate(Dataset = replace(Dataset, Dataset == "MT_nM_41", "Omicron"))%>%# Change col name for visualisation
  mutate(Dataset = factor(Dataset, levels = c("Wuhan","Omicron")))%>%
        ggplot(aes(x=Aff, color=Dataset))+stat_ecdf(size=1)+theme_pubr(base_size = 22)+
        grids()+ guides(color = guide_legend(nrow = 2))+ylab("Cumulative Freq. of Peptides")+
        xlab("Binding Affinity\n(nM)")+geom_vline(xintercept = 500,linetype="dashed",alpha=0.5)+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22))+
       theme(strip.text.x = element_text(size = 22))

# ECDF comparing BA1 Wuhan vs Mutated: same as above but log10 x axis
ECDF_NM_LOG10_PLT=SPIKE_AGRETOPICITY %>% select(Peptide, nM_41, MT_nM_41)%>%
  pivot_longer(cols = c(nM_41,MT_nM_41), names_to = "Dataset",values_to = "Aff")%>%
  mutate(Dataset = replace(Dataset, Dataset == "nM_41", "Wuhan"))%>%
  mutate(Dataset = replace(Dataset, Dataset == "MT_nM_41", "Omicron"))%>%
  mutate(Dataset = factor(Dataset, levels = c("Wuhan","Omicron")))%>%
        ggplot(aes(x=Aff, color=Dataset))+stat_ecdf(size=1)+theme_pubr(base_size = 22)+
        grids()+ guides(color = guide_legend(nrow = 2))+ylab("Cumulative Freq. of Peptides")+
        xlab("Log10 Binding Affinity\n(nM)")+geom_vline(xintercept = 500,linetype="dashed",alpha=0.5)+
        scale_x_log10()+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22))

# Violin plot: same data (log 10 y axis)
VIOLIN_NM_PLT=SPIKE_AGRETOPICITY %>% select(Peptide, nM_41, MT_nM_41)%>%
  pivot_longer(cols = c(nM_41,MT_nM_41), names_to = "Dataset",values_to = "Aff") %>%
  mutate(Dataset = replace(Dataset, Dataset == "nM_41", "Wuhan"))%>%
  mutate(Dataset = replace(Dataset, Dataset == "MT_nM_41", "Omicron"))%>%
  mutate(Dataset = factor(Dataset, levels = c("Wuhan","Omicron")))%>%
         ggviolin(x="Dataset",y="Aff",color = "Dataset",add="boxplot",trim=TRUE,linewidth=2)+
        scale_y_log10()+theme_pubr(base_size = 22)+stat_compare_means(label="p.signif",label.x.npc = "center")+
        ylab("Log10 Binding Affinity\n(nM)")+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22))

# Violin plot: same data (log 10 y axis)
VIOLIN_NM_PLT=SPIKE_AGRETOPICITY %>% select(Peptide, nM_41, MT_nM_41)%>%
  pivot_longer(cols = c(nM_41,MT_nM_41), names_to = "Dataset",values_to = "Aff") %>%
  mutate(Dataset = replace(Dataset, Dataset == "nM_41", "Wuhan"))%>%
  mutate(Dataset = replace(Dataset, Dataset == "MT_nM_41", "Omicron"))%>%
  mutate(Dataset = factor(Dataset, levels = c("Wuhan","Omicron")))%>%
        ggplot(aes(x=Dataset, y=Aff,color=Dataset))+ geom_violin(lwd=1.0)+ geom_boxplot(width=0.15)+
        scale_y_log10()+theme_pubr(base_size = 22)+stat_compare_means(label="p.signif",label.x.npc = "center")+
        ylab("Log10 Binding Affinity\n(nM)")+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22))


```

# BA1: SUPERTYPES ASSOCIATION
- Supplementary Fig
- HLA-A/B and C are produced separately. A/B have defined supertype groups. C is less clear, therefore we just take the prefix i.e HLA-C01, or C03, etc.
```{r,dpi=300, fig.width  = 11, fig.height = 9}
AB_SUPERTYPES=fread("HLA_AB_SUPERTYPES.csv") %>% mutate(Allele = gsub("\\*","",Allele))%>% mutate(Allele = paste0("HLA-",Allele))%>% dplyr::rename(HLA_Allele = Allele)
# Curate data for analysis: spike only
DATA_FOR_ANALYSIS=FULL_MUTATIONS_DT %>% filter(Protein == 'surface glycoprotein')%>% # filter spike
  select(Peptide, VariantAlignment, Predicted_Binding, nM_41, MT_nM_41, Binder,MT_Binder) %>%  # select rewquired columns
  separate_rows_(cols = c("Predicted_Binding","nM_41","MT_nM_41","Binder","MT_Binder"),sep=",")%>% # separate collapsed rows
  mutate(nM_41 = as.numeric(nM_41), MT_nM_41=as.numeric(MT_nM_41))%>%
  mutate(Agretopicity = MT_nM_41/nM_41)

# Link HLA alleles in our data with HLA A/B supertypes
HLA_A_B_AGRETOPICITY=DATA_FOR_ANALYSIS %>% dplyr::rename(HLA_Allele=Predicted_Binding)%>%
  filter(HLA_Allele %in% grep("HLA-A|HLA-B",HLA_Allele,value = T)) %>%
  inner_join(AB_SUPERTYPES) %>%
  filter(! (Binder == "NONBINDER" & MT_Binder == 'NONBINDER')) # Exclude peptide-MHC samples where both WT and MT don't bind MHC k. Irrelevant obs
# Grep the prefix and use as supertype for HLA-C alleles.
HLA_CDATASET_AGRETOPICITY=DATA_FOR_ANALYSIS %>% dplyr::rename(HLA_Allele=Predicted_Binding)%>%
  filter(!HLA_Allele %in% grep("HLA-A|HLA-B",HLA_Allele,value = T))%>%
  mutate(Supertype = str_extract(HLA_Allele,"HLA-C[0-9][0-9]"))%>%
  mutate(Supertype = gsub("HLA\\-","",Supertype))%>%
  filter(! (Binder == "NONBINDER" & MT_Binder == 'NONBINDER')) # Exclude peptide-MHC samples where both WT and MT don't bind MHC k. Irrelevant obs

```



# PRARR plots
- FIG 2D
-
```{r}

DATA_FOR_SUPERTYPE_MUTPOSHLA=FULL_MUTATIONS_DT %>% mutate(Length = nchar(Peptide)) %>%
  separate_rows_(cols="MutationPos", sep=",") %>% #Create a row for each mutation position
  select(Peptide, VariantAlignment, MutationPos, Predicted_Binding, Binder, MT_Binder, nM_41, MT_nM_41,Mutation)%>%
  separate_rows_(cols = "Mutation",sep=",")%>% # Seperate each mutatiopn
  separate_rows_(cols = c("Predicted_Binding","Binder","MT_Binder","nM_41","MT_nM_41"), sep=",")%>%
  mutate(MutationPos = as.numeric(MutationPos))%>%
  filter(! (Binder == "NONBINDER" & MT_Binder == "NONBINDER"))%>% # Exclude obs where both WT and MT dot bind MHC
  mutate(nM_41 = as.numeric(nM_41), MT_nM_41=as.numeric(MT_nM_41))%>%
  mutate(Agretopicity = MT_nM_41/nM_41)%>% dplyr::rename(HLA_Allele=Predicted_Binding)

#HLA C data, extract prefix and use as supertype
DATA_FOR_SUPERTYPE_MUTPOSHLA_C=DATA_FOR_SUPERTYPE_MUTPOSHLA%>%
  filter(!HLA_Allele %in% grep("HLA-A|HLA-B",HLA_Allele,value = T))%>%
  mutate(Supertype = str_extract(HLA_Allele,"HLA-C[0-9][0-9]"))
#HLA0B data
DATA_FOR_SUPERTYPE_MUTPOSHLA_A_B=DATA_FOR_SUPERTYPE_MUTPOSHLA%>%
  filter(HLA_Allele %in% grep("HLA-A|HLA-B",HLA_Allele,value = T)) %>% inner_join(AB_SUPERTYPES)

```

```{r,dpi=300, fig.width = 17}

DATA_FOR_SUPERTYPE_MUTPOSHLA_A_B %>% rbind(DATA_FOR_SUPERTYPE_MUTPOSHLA_C)%>%
  filter(! Mutation == "NA")%>% filter(Supertype == "B07")%>%  # Filter for B07 ligands to examinea agretopicity of different mutaytions
  mutate(MT_Binder = factor(MT_Binder, levels = c("NONBINDER","BINDER")))%>%
  arrange(desc(Agretopicity))%>%select(Peptide, VariantAlignment, HLA_Allele, MT_Binder, Mutation, Agretopicity)%>%
  distinct %>% DT::datatable()

PRARR_PER_MUT_PLT=DATA_FOR_SUPERTYPE_MUTPOSHLA_A_B %>% rbind(DATA_FOR_SUPERTYPE_MUTPOSHLA_C)%>%
  filter(! Mutation == "NA")%>% filter(Supertype == "B07")%>%  # FILTER B07 LIGANDS
  mutate(MT_Binder = factor(MT_Binder, levels = c("NONBINDER","BINDER")))%>%
  arrange(desc(Agretopicity))%>%select(Peptide, VariantAlignment, HLA_Allele, MT_Binder, Mutation, Agretopicity)%>%
  distinct %>% mutate(pMHC = paste0(Peptide,"_", gsub("HLA-","",HLA_Allele)))%>%#Create pMHC column
  filter(! Mutation %in% c('Q493R','Q498R','K856R','N501Y','G142D','E484A','V143-', #exclude mutations with one observation
                           'Y144-','Y505H', "E484A","G142D","K856R","N501Y","Q493F","Q498F","Y505H"))%>%
  ggbarplot(x="pMHC",y="Agretopicity",fill="MT_Binder",position = position_dodge2())+ # visualise tyhe agretopicity grouped by mutation
  theme_pubr(base_size = 20)+facet_grid(~Mutation,scales="free",space="free")+scale_y_log10()+
  rotate_x_text()+ylab("Agretopicity (log10)")+
  theme(strip.background = element_blank())+theme(strip.text = element_text(size=22))+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22))+
       theme(strip.text.x = element_text(size = 22))
PRARR_PER_MUT_PLT

```


```{r,dpi=300}


FULL_MUTATIONS_DT %>% filter(Peptide %in% grep("PRRA", Peptide, value=T))%>%
    select(Peptide, VariantAlignment, Predicted_Binding, nM_41, MT_nM_41, Binder, MT_Binder)%>%
    separate_rows_(cols = c("Predicted_Binding", "nM_41", "MT_nM_41", "Binder", "MT_Binder"),sep=",")%>%
    filter(! (Binder == "NONBINDER" & MT_Binder == "NONBINDER"))%>%mutate(nM_41=as.numeric(nM_41))%>% mutate(MT_nM_41=as.numeric(MT_nM_41))%>%
    mutate(Agretopicity = MT_nM_41/nM_41)%>% mutate(pMHC = paste0(Peptide,"_", gsub("HLA-","",Predicted_Binding)))%>%
    mutate(MT_Binder = factor(MT_Binder, levels = c("NONBINDER","BINDER")))%>%
    arrange(desc(Agretopicity))%>%
    dplyr::rename("Variant Peptide"=VariantAlignment, "HLA Allele"=Predicted_Binding, WT_BA=nM_41, MT_BA=MT_nM_41)%>%DT::datatable()

PRRAR_PMHC_PLT=FULL_MUTATIONS_DT %>% filter(Peptide %in% grep("PRRA", Peptide, value=T))%>% #Filterr for peptides with PRRA motif
    select(Peptide, VariantAlignment, Predicted_Binding, nM_41, MT_nM_41, Binder, MT_Binder)%>%
    separate_rows_(cols = c("Predicted_Binding", "nM_41", "MT_nM_41", "Binder", "MT_Binder"),sep=",")%>%
    filter(! (Binder == "NONBINDER" & MT_Binder == "NONBINDER"))%>%mutate(nM_41=as.numeric(nM_41))%>% # exclude MHC where WT/mt not predicted to bind
    mutate(MT_nM_41=as.numeric(MT_nM_41))%>% mutate(Agretopicity = MT_nM_41/nM_41)%>%
    mutate(pMHC = paste0(Peptide,"_", gsub("HLA-","",Predicted_Binding)))%>% # create pMHC column
    mutate(MT_Binder = factor(MT_Binder, levels = c("NONBINDER","BINDER")))%>%
    arrange(desc(Agretopicity))%>%
    ggbarplot(x="pMHC",y="Agretopicity",fill="MT_Binder")+
    theme_pubr(base_size = 22)+rotate_x_text()+scale_y_log10()+ylab("Agretopicity (log10)")+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=20))


```

# FIG 2F

```{r, fig.width = 10,dpi=300}
# How many HLA does WT vs MT peptide bind?
NUM_HLAS_BIND_PEPS_PLT=FULL_MUTATIONS_DT %>% select(Peptide,Binder, Predicted_Binding, MT_Binder)%>%
  separate_rows_(cols = c("Binder","Predicted_Binding","MT_Binder"),sep=",")%>%
  filter(! (Binder == "NONBINDER" & MT_Binder == "NONBINDER"))%>% # Exclde nonbinders
  pivot_longer(cols = c(Binder,MT_Binder))%>% group_by(Peptide, name,value)%>%
  dplyr::summarise(n=n())%>% filter(! value == "NONBINDER")%>% arrange(desc(name), desc(n))%>% # Count by group, then exclde nonbinders.. i.e we're only itnrested in how many binders per HLA
  dplyr::rename(Dataset=name)%>%mutate(Dataset = replace(Dataset, Dataset == "MT_Binder", "Omicron_Binders"))%>%
  mutate(Dataset = replace(Dataset, Dataset == "Binder", "Wuhan_Binders"))%>% # clean names
  mutate(Dataset = factor(Dataset, levels = c("Wuhan_Binders","Omicron_Binders")))%>% # Plot, dodged per omicron vs wuhan
  ggbarplot(x="Peptide",y="n", fill="Dataset",position = position_dodge())+theme_pubr(base_size = 22)+rotate_x_text()+ylab("# HLAs predicted to bind")+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22))


```



```{r}
# Read in subvariant plots from 3_Analyuse_AgPRes_Subvariants.rmd
SUBVAR_ECDF_PLT=readRDS("SUBVAR_ECDF_PLT.rds")+ylab("Cumulative Freq. of Peptides")+theme_pubr(base_size = 22)+theme(strip.text.x = element_text(size = 22))+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22))
SUBVAR_VIOLIN_PLT = readRDS("SUBVAR_VIOLIN_PLT.rds")+theme_pubr(base_size = 22)+theme(strip.text.x = element_text(size = 22))+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22))
SUBVAR_SUPERTYPE_BOXPLT = readRDS("SUBVAR_SUPERTYPE_BOXPLT.rds")+theme_pubr(base_size = 22)+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22))+theme(strip.text.x = element_text(size = 22))

BA1_SUPERTYPE_PLT = HLA_A_B_AGRETOPICITY %>% rbind(HLA_CDATASET_AGRETOPICITY)%>% dplyr::rename(Wuhan=nM_41, Omicron=MT_nM_41) %>% filter(Supertype %in% c("A02","A03","B07","C01"))%>%
        ggpaired(cond1 = "Wuhan", cond2 = "Omicron",line.color = "gray")+theme_pubr(base_size = 20)+facet_wrap(~Supertype,ncol=9)+scale_y_log10()+stat_compare_means(label = "p.signif",label.x.npc = "center")+rotate_x_text(angle=45)+ylab("Binding Affinity (nM)")+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22))

```

# Main


```{r,dpi=300, fig.height = 42, fig.width = 34}
A_GRID=cowplot::plot_grid(ECDF_NM_PLT, ECDF_NM_LOG10_PLT, VIOLIN_NM_PLT, align="hv",nrow=1)

B_GRID_i=cowplot::plot_grid(SUBVAR_ECDF_PLT+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22)),SUBVAR_VIOLIN_PLT+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22)), align="hv",nrow=1,axis="blt")
B_GRID_ii=cowplot::plot_grid(SUBVAR_SUPERTYPE_BOXPLT+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22)), align="hv",nrow=1,axis="blt")

C_GRID = cowplot::plot_grid(PRARR_PER_MUT_PLT, nrow=1, align="hv", axis="bl")
D_F_GRID = cowplot::plot_grid( PRRAR_PMHC_PLT , nrow=1,align="hv",axis="bl")
E_GRID = cowplot::plot_grid( PRRAR_PMHC_PLT+labs(fill=""),NUM_HLAS_BIND_PEPS_PLT+
  theme(legend.text = element_text(size=22))+theme(legend.title = element_text(size=22)) , nrow=1, rel_widths = c(0.20,0.8))

cowplot::plot_grid(A_GRID, B_GRID_i,B_GRID_ii, C_GRID, E_GRID, nrow =5, rel_heights = c(0.8,0.7,1.7,1.1,0.9), align="v", axis="blt")

```


# Supplementary
## BA1 SPIKE

```{r,dpi=300, fig.width = 8,fig.height=6}

AGRETOPICITY_DT=FULL_MUTATIONS_DT %>% select(Peptide, VariantAlignment, Protein, Predicted_Binding, BA_Rank, MT_BA_Rank,Binder,MT_Binder) %>%
        separate_rows_(cols = c("Predicted_Binding","BA_Rank","MT_BA_Rank","Binder","MT_Binder"),sep=",") %>% mutate(BA_Rank = as.numeric(BA_Rank)) %>% mutate(MT_BA_Rank = as.numeric(MT_BA_Rank))

SPIKE_AGRETOPICITY=AGRETOPICITY_DT %>% filter(Protein == 'surface glycoprotein') %>% mutate(Length = nchar(Peptide))

SPIKE_AGRETOPICITY=SPIKE_AGRETOPICITY %>% filter(! (Binder == "NONBINDER" & MT_Binder == "NONBINDER"))

SPIKE_ECDF_NM_PLT=SPIKE_AGRETOPICITY %>% select(Peptide, BA_Rank, MT_BA_Rank)%>% pivot_longer(cols = c(BA_Rank, MT_BA_Rank), names_to = "Dataset",values_to = "Aff")%>%
  mutate(Dataset = replace(Dataset, Dataset == "BA_Rank", "Wuhan"))%>%mutate(Dataset = replace(Dataset, Dataset == "MT_BA_Rank", "Omicron"))%>%
  mutate(Dataset = factor(Dataset, levels = c("Wuhan","Omicron")))%>%
        ggplot(aes(x=Aff, color=Dataset))+stat_ecdf(size=1)+theme_pubr(base_size = 20)+grids()+ guides(color = guide_legend(nrow = 2))+ylab("Cumulative Freq. of Peptides")+xlab("Binding Affinity\n(Rank)")+geom_vline(xintercept = 2,linetype="dashed",alpha=0.5)


SPIKE_ECDF_NM_LOG10_PLT=SPIKE_AGRETOPICITY %>% select(Peptide, BA_Rank, MT_BA_Rank)%>% pivot_longer(cols = c(BA_Rank, MT_BA_Rank), names_to = "Dataset",values_to = "Aff")%>%
  mutate(Dataset = replace(Dataset, Dataset == "BA_Rank", "Wuhan"))%>%mutate(Dataset = replace(Dataset, Dataset == "MT_BA_Rank", "Omicron"))%>%
  mutate(Dataset = factor(Dataset, levels = c("Wuhan","Omicron")))%>%
        ggplot(aes(x=Aff, color=Dataset))+stat_ecdf(size=1)+theme_pubr(base_size = 20)+grids()+ guides(color = guide_legend(nrow = 2))+ylab("Cumulative Freq. of Peptides")+xlab("Log10 Binding Affinity\n(Rank)")+geom_vline(xintercept = 2,linetype="dashed",alpha=0.5)+scale_x_log10()


SPIKE_VIOLIN_NM_PLT=SPIKE_AGRETOPICITY %>% select(Peptide, BA_Rank, MT_BA_Rank)%>% pivot_longer(cols = c(BA_Rank, MT_BA_Rank), names_to = "Dataset",values_to = "Aff") %>%
  mutate(Dataset = replace(Dataset, Dataset == "BA_Rank", "Wuhan"))%>%mutate(Dataset = replace(Dataset, Dataset == "MT_BA_Rank", "Omicron"))%>%
  mutate(Dataset = factor(Dataset, levels = c("Wuhan","Omicron")))%>%
         ggviolin(x="Dataset",y="Aff",color = "Dataset",add="boxplot",trim=TRUE)+scale_y_log10()+theme_pubr(base_size = 20)+stat_compare_means(label="p.signif",label.x.npc = "center")+ylab("Binding Affinity\n(Rank)")+geom_hline(yintercept = 2,linetype="dashed",alpha=0.5)


```


## BA1 ALL PROTEINS

```{r,dpi=300, fig.width = 8,fig.height=6}

AGRETOPICITY_DT=FULL_MUTATIONS_DT %>% select(Peptide, VariantAlignment, Protein, Predicted_Binding, BA_Rank, MT_BA_Rank,Binder,MT_Binder) %>%
        separate_rows_(cols = c("Predicted_Binding","BA_Rank","MT_BA_Rank","Binder","MT_Binder"),sep=",") %>% mutate(BA_Rank = as.numeric(BA_Rank)) %>% mutate(MT_BA_Rank = as.numeric(MT_BA_Rank))


AGRETOPICITY_DT=AGRETOPICITY_DT %>% filter(! (Binder == "NONBINDER" & MT_Binder == "NONBINDER"))

ECDF_NM_PLT=AGRETOPICITY_DT %>% select(Peptide, BA_Rank, MT_BA_Rank)%>% pivot_longer(cols = c(BA_Rank, MT_BA_Rank), names_to = "Dataset",values_to = "Aff")%>%
  mutate(Dataset = replace(Dataset, Dataset == "BA_Rank", "Wuhan"))%>%mutate(Dataset = replace(Dataset, Dataset == "MT_BA_Rank", "Omicron"))%>%
  mutate(Dataset = factor(Dataset, levels = c("Wuhan","Omicron")))%>%
        ggplot(aes(x=Aff, color=Dataset))+stat_ecdf(size=1)+theme_pubr(base_size = 20)+grids()+ guides(color = guide_legend(nrow = 2))+ylab("Cumulative Freq. of Peptides")+xlab("Binding Affinity\n(Rank)")+geom_vline(xintercept = 2,linetype="dashed",alpha=0.5)

ECDF_NM_LOG10_PLT=AGRETOPICITY_DT %>% select(Peptide, BA_Rank, MT_BA_Rank)%>% pivot_longer(cols = c(BA_Rank, MT_BA_Rank), names_to = "Dataset",values_to = "Aff")%>%
  mutate(Dataset = replace(Dataset, Dataset == "BA_Rank", "Wuhan"))%>%mutate(Dataset = replace(Dataset, Dataset == "MT_BA_Rank", "Omicron"))%>%
  mutate(Dataset = factor(Dataset, levels = c("Wuhan","Omicron")))%>%
        ggplot(aes(x=Aff, color=Dataset))+stat_ecdf(size=1)+theme_pubr(base_size = 20)+grids()+ guides(color = guide_legend(nrow = 2))+ylab("Cumulative Freq. of Peptides")+xlab("Log10 Binding Affinity\n(Rank)")+geom_vline(xintercept = 2,linetype="dashed",alpha=0.5)+scale_x_log10()

VIOLIN_NM_PLT=AGRETOPICITY_DT %>% select(Peptide, BA_Rank, MT_BA_Rank)%>% pivot_longer(cols = c(BA_Rank, MT_BA_Rank), names_to = "Dataset",values_to = "Aff") %>%
  mutate(Dataset = replace(Dataset, Dataset == "BA_Rank", "Wuhan"))%>%mutate(Dataset = replace(Dataset, Dataset == "MT_BA_Rank", "Omicron"))%>%
  mutate(Dataset = factor(Dataset, levels = c("Wuhan","Omicron")))%>%
         ggviolin(x="Dataset",y="Aff",color = "Dataset",add="boxplot",trim=TRUE)+scale_y_log10()+theme_pubr(base_size = 20)+stat_compare_means(label="p.signif",label.x.npc = "center")+ylab("Binding Affinity\n(Rank)")+geom_hline(yintercept = 2,linetype="dashed",alpha=0.5)

```
## subvar spike
```{r}
SUBVAR_VIOLIN_PLT_RANK=readRDS( "SUBVAR_RANK_VIOLIN_PLT.rds")

SUBVAR_ECDF_PLT_RANK=readRDS( "SUBVAR_RANK_ECDF_PLT.rds")
SUBVAR_ECDF_PLT_RANK=SUBVAR_ECDF_PLT_RANK+xlab("Log 10 Binding Affinity\n(Rank)")
```

# all proteins subvar

```{r}
SUBVAR_VIOLIN_ALLPROT_PLT = readRDS( "SUBVAR_RANK_ALL_VIOLIN_PLT.rds")

SUBVAR_ECDF_ALL_PLT= readRDS( "SUBVAR_RANK_ALL_ECDF_PLT.rds")
SUBVAR_ECDF_ALL_PLT=SUBVAR_ECDF_ALL_PLT+xlab("Log 10 Binding Affinity\n(Rank)")
```



```{r, fig.width = 18, fig.height = 18}
SPIKE_PAIRED_EPITOPE_PLT=FULL_MUTATIONS_DT %>% select(Peptide, VariantAlignment, Predicted_Binding, Binder,MT_Binder, BA_Rank, MT_BA_Rank, Protein) %>% separate_rows_(c("Predicted_Binding","Binder","BA_Rank","MT_Binder","MT_BA_Rank","Protein"),sep=",")%>%ungroup()%>%
        filter(! (Binder == "NONBINDER" & MT_Binder == "NONBINDER")) %>% mutate(BA_Rank = as.numeric(BA_Rank))%>% mutate(MT_BA_Rank = as.numeric(MT_BA_Rank))%>%filter(Protein == 'surface glycoprotein')%>%dplyr::rename(Wuhan=BA_Rank, Omicron=MT_BA_Rank)%>%
  ggpaired(cond1 = "Wuhan", cond2 = "Omicron")+facet_wrap(~Peptide)+theme_pubr(base_size = 20)+stat_compare_means(label = "p.signif",label.x.npc = "center")+scale_y_log10()+ggtitle("Spike")+ylab("BA Rank")

```

```{r, fig.width = 9, fig.height = 9}
NON_SPIKE_PAIRED_EPITOPE_PLT=FULL_MUTATIONS_DT %>% select(Peptide, VariantAlignment, Predicted_Binding, Binder,MT_Binder, BA_Rank, MT_BA_Rank, Protein) %>% separate_rows_(c("Predicted_Binding","Binder","BA_Rank","MT_Binder","MT_BA_Rank","Protein"),sep=",")%>%ungroup()%>%
        filter(! (Binder == "NONBINDER" & MT_Binder == "NONBINDER")) %>% mutate(BA_Rank = as.numeric(BA_Rank))%>%
  mutate(MT_BA_Rank = as.numeric(MT_BA_Rank))%>%filter(!Protein == 'surface glycoprotein')%>%
  dplyr::rename(Wuhan=BA_Rank, Omicron=MT_BA_Rank)%>%
  ggpaired(cond1 = "Wuhan", cond2 = "Omicron")+facet_wrap(~Peptide,ncol=8)+theme_pubr(base_size = 20)+stat_compare_means(label = "p.signif",label.x.npc = "center")+scale_y_log10()+ggtitle("Non-spike Proteins")+ylab("BA Rank")

```

#BA1 SUPERTYPES

```{r}
BA1_SUPERTYPE_BOXPLT=readRDS(file="BA1_SUPERTYPE_BOXPLT.rds")

```

# Supp Fig: Page 1

```{r,dpi=300, fig.width = 24, fig.height = 6}

GRID_1 = cowplot::plot_grid(SPIKE_ECDF_NM_PLT, SPIKE_ECDF_NM_LOG10_PLT, SPIKE_VIOLIN_NM_PLT, nrow=1)
GRID_2 = cowplot::plot_grid(ECDF_NM_PLT, ECDF_NM_LOG10_PLT, VIOLIN_NM_PLT, nrow=1)
GRID_3 = cowplot::plot_grid(SUBVAR_ECDF_PLT_RANK,SUBVAR_VIOLIN_PLT_RANK, nrow=1)
GRID_4 = cowplot::plot_grid(SUBVAR_ECDF_ALL_PLT, SUBVAR_VIOLIN_ALLPROT_PLT, nrow=1)
GRID_5 = cowplot::plot_grid(BA1_SUPERTYPE_BOXPLT, nrow=1)
```

```{r,dpi=300, fig.width = 24, fig.height = 24}

cowplot::plot_grid(GRID_1,GRID_2,GRID_3, GRID_4,GRID_5,align="hv",axis="l",nrow=5)

```


## Per a reviewer's suggestion: plot variant-specific seqlogos for HLA supertypes of interest

```{r}
library("ggseqlogo")

```

```{r}
### filter 9mers
LOGO_PLT=SUBVAR_SUPERTYPE_BOXPLT$data %>% filter(nchar(Peptide)==9) %>%
    mutate(Group_logo= paste0(Variant,"_",Supertype,"_",condition))%>% arrange(Group_logo)
MT_LOGO=LOGO_PLT %>% filter(condition=="Omicron")%>% select(!Peptide) %>% dplyr::rename(Peptide=VariantAlignment)%>% distinct() # seperate Omicron and Wuhan
WT_LOGO=LOGO_PLT %>% filter(condition=="Wuhan")%>% select(!VariantAlignment)%>% distinct()

LOGO_LONG = MT_LOGO %>% rbind(WT_LOGO)%>% arrange((condition),Supertype, Variant)%>% filter(Supertype %in% c("A02","A03","B07"))

```


```{r,dpi=300, fig.height = 18, fig.width = 20}

logo_lists=by(LOGO_LONG,LOGO_LONG$Group_logo,function(x) setNames(x$Peptide,x$name))
logo_lists = split(x=as.character(LOGO_LONG$Peptide), f=LOGO_LONG$Group_logo)
ggseqlogo::ggseqlogo(logo_lists, ncol=4)
```



## page 3
```{r,dpi=300, fig.width = 24, fig.height = 33}
EMPTY=NULL

cowplot::plot_grid(SPIKE_PAIRED_EPITOPE_PLT, NON_SPIKE_PAIRED_EPITOPE_PLT, nrow=2, rel_heights = c(1.3,0.7))

```








